#code for downloading sequences from genbank and cleaning up
#example code from 
#http://www.jcsantosresearch.org/Class_2014_Spring_Comparative/pdf/week_2/Jan_13_15_2015_GenBank_part_2.pdf

setwd('~/jsb439@cornell.edu/amelia-bivalves/tubeworms')

require(ape)
require(seqinr)
require(plyr)
require(insect)
require(readxl)
require(phytools)
require(Rphylopars)
require(nlme)
require(geiger)
require(ggplot2)
require(gridExtra)
require(olsrr)
require(WGCNA)
require(evomap)


#setwd('~/jsb439@cornell.edu/amelia-bivalves')
source(file="/Users/cotinga/jsb439@cornell.edu/Code/deep_sea_evolution/functions.R")

#set up CO1
{
#read in the excel spreadsheet
data<-read_excel('~/jsb439@cornell.edu/amelia-bivalves/tubeworms/Tubeworms2.xlsx', sheet=2)


#extract only ingroup 
ingroup <- data[data$is.ingroup=='1',]
ingroup_CO1_accession_numbers <- ingroup$CO1
CO1_accession_numbers <- ingroup_CO1_accession_numbers[!is.na(ingroup_CO1_accession_numbers)] #filter out NAs
ingroup_pruned <- ingroup[!is.na(ingroup$CO1),] #filter out rows without CO1 accession numbers (for species labels)

#extract only  (outgroup)
outgroup <- data[data$is.outgroup==1,]
outgroup <- outgroup[!is.na(outgroup$CO1),]

outgroup_CO1_accession_numbers <- outgroup$CO1 #extract only CO1 accession numbers

CO1<-read.GenBank(c(CO1_accession_numbers, outgroup_CO1_accession_numbers), quiet=F)
SequenceNames<-paste(names(CO1), attr(CO1, "species"), sep ="_CO1_") #generate new vector of taxon IDs

#clean names
SequenceNames<-gsub(pattern = ".", "", SequenceNames,fixed = T)
SequenceNames<-gsub(pattern = "'", "", SequenceNames,fixed = T)
SequenceNames<-gsub(pattern = "-", "_", SequenceNames,fixed = T)

names(CO1) <- SequenceNames #assign new names

#write unaligned fasta out for alignment in MAAFT and then analysis in RAxML
#write.dna(CO1, format='fasta', file='unalignedtubeworms_2_19_22.fas')
}

## read in the tree generated by IQTREE, swap names
{
CO1_ML.unconstrained<-read.tree('~/jsb439@cornell.edu/amelia-bivalves/tubeworms/IQTREE/unconstrained/unconstrained.treefile')
# swaps<-c("AB088673_CO1_cf._Alaysia_sp._A4", "AB073491_CO1_Arcovestia_ivanovi", "AY129124_CO1_Lamellibrachia_cf._luymesi",
#                                   "AY129132_CO1_Lamellibrachia_cf._luymesi", "EU046616_CO1_Lamellibrachia_anaximandri", "HQ154525_CO1_Lamellibrachia_sp._Lam1", 
#                                   "HQ154526_CO1_Lamellibrachia_sp._Lam2", "HQ396891_CO1_Lamellibrachia_sp._DBUA_01146", "MH670807_CO1_Lamellibrachia_donwalshi", 
#                                   "KJ566961_CO1_Lamellibrachia_sp._2_SP-2014", "D38030_CO1_Lamellibrachia_satsuma", "U74078_CO1_Basibranchia_sp.__Mariana_Trough_", 
#                                   "U74055_CO1_Lamellibrachia_barhami", "D38029_CO1_Lamellibrachia_sp.", "LC064365_CO1_Lamellibrachia_sp._GK-2015", 
#                                   "D50592_CO1_Lamellibrachia_sp.", "U74061_CO1_Lamellibrachia_columna", "AB055209_CO1_Lamellibrachia_sp._L4", 
#                                   "AB088674_CO1_Lamellibrachia_sp._L6", "AB055210_CO1_Lamellibrachia_sp._L5", "AB088675_CO1_Lamellibrachia_sp._L7", 
#                                   "AB264601_CO1_Lamellibrachia_juni", "AB242857_CO1_Oasisia_fujikurai", "FJ667533_CO1_Oasisia_sp._SBJ-2009", 
#                                   "U74069_CO1_Oasisia_alvinae", "FJ667534_CO1_Oasisia_sp._SBJ-2009", "FJ667535_CO1_Oasisia_sp._SBJ-2009", 
#                                   "U74056_CO1_Ridgeia_sp.__Southern_Explorer_Ridge_", "U74073_CO1_Ridgeia_piscesae", "U74075_CO1_Tevnia_jerichonana", 
#                                   "U74074_CO1_Riftia_pachyptila", "FM178480_CO1_Sclerolinum_contortum", "U74068_CO1__Loihi_Seamount__perviate_pogonophoran", 
#                                   "FJ347644_CO1_Sclerolinum_brattstromi", "FJ480347_CO1_Siboglinum_sp._1_SBJ-2009", "FJ480378_CO1_Siboglinum_sp._2_SBJ-2009", 
#                                   "FJ480394_CO1_Siboglinum_sp._3_SBJ-2009", "FJ480396_CO1_Siboglinum_sp._4_SBJ-2009", "FJ480358_CO1_Siboglinum_sp._5_SBJ-2009", 
#                                   "FJ480364_CO1_Siboglinum_sp._6_SBJ-2009", "FJ480360_CO1_Sibloglinidae_cf._Cyclobrachia_sp._SBJ-2009", 
#                                   "FJ480399_CO1_Siboglinum_poseidoni", "KJ789169_CO1_Siboglinum_ekmani", "FJ480389_CO1_Polybrachia_sp._2_SBJ-2009", 
#                                   "FJ480391_CO1_Polybrachia_sp._3_SBJ-2009", "FJ480393_CO1_Polybrachia_sp._1_SBJ-2009", "KJ789162_CO1_Galathealinum_brachiosum", 
#                                   "U74066_CO1_Galathealinum_brachiosum", "FJ480357_CO1_Bobmarleya_gadensis", "FJ480370_CO1_Spirobrachia_tripeira", 
#                                   "FJ483547_CO1_Spirobrachia_cf._grandis_SJ-2009", "KJ789171_CO1_Spirobrachia_sp._YL-2014", "FJ480376_CO1_Lamellisabella_denticulata", 
#                                   "KJ789170_CO1_Siboglinum_fiordicum", "FM178481_CO1_Oligobrachia_haakonmosbiensis", "MH619675_CO1_Oligobrachia_sp._1125-3", 
#                                   "FJ480388_CO1_Lamellisabella_sp._SBJ-2009", "AB259569_CO1_Osedax_japonicus", "DQ996629_CO1_Osedax_sp._3_SBJ-2006", 
#                                   "EU223340_CO1_Osedax_sp.__orange-collar_", "AY827562_CO1_Osedax_mucofloris", "FJ347615_CO1_Osedax_sp.__white-collar_", 
#                                   "FM998109_CO1_Osedax_sp._sagami-7", "JX280613_CO1_Osedax_sp.__MB16_", "AY586495_CO1_Osedax_frankpressi", 
#                                   "LC106303_CO1_Osedax_sp._1336_61_2", "LC381421_CO1_Osedax_braziliensis", "EU223307_CO1_Osedax_rubiplumus", 
#                                   "FM998082_CO1_Osedax_sp._sagami-4", "FM998081_CO1_Osedax_sp._sagami-3", "JF509949_CO1_Osedax_roseus", 
#                                   "JX280609_CO1_Osedax_sp.__MB17_", "EU223356_CO1_Osedax_sp.__nude-palp-A_", "EU236218_CO1_Osedax_sp.__nude-palp-B_", 
#                                   "KP119563_CO1_Osedax_sp._Nude-palp-G", "FJ347643_CO1_Osedax_sp.__nude-palp-F_", "KJ598040_CO1_Osedax_rogersi", 
#                                   "KF444422_CO1_Osedax_antarcticus", "KJ598038_CO1_Osedax_crouchi", "KJ598039_CO1_Osedax_nordenskjoeldi", 
#                                   "EU267675_CO1_Osedax_sp.__nude-palp-C_", "FM998107_CO1_Osedax_sp._sagami-6", "FJ347630_CO1_Osedax_sp.__nude-palp-D_", 
#                                   "FM998110_CO1_Osedax_sp._sagami-8", "FJ347632_CO1_Osedax_sp.__nude-palp-E_", "FJ347639_CO1_Osedax_sp.__green-palp_", 
#                                   "FM998083_CO1_Osedax_sp._sagami-5", "KT860548_CO1_Osedax_sp.__Mediterranea_", "KP119564_CO1_Osedax_priapus", 
#                                   "FJ347616_CO1_Osedax_sp.__yellow-patch_", "KT860545_CO1_Osedax_deceptionensis", "DQ996622_CO1_Osedax_sp._1_SBJ-2006", 
#                                   "JF317197_CO1_Hesiospina_aurantiaca", "MH202758_CO1_Amphisamytha_sp._n._1_YZ-2018", "MH202759_CO1_Amphisamytha_sp._n._2_YZ-2018", 
#                                   "KP144926_CO1_Nephtys_hombergii", "AY428839_CO1_Owenia_fusiformis", "MH202756_CO1_Hesiolyra_cf._bergi_YZ-2018", 
#                                   "AF317287_CO1_Seepiophila_jonesi", "AY129134_CO1_Escarpia_sp._GB425-GoM", "AY326303_CO1_Escarpia_southwardae", 
#                                   "U74065_CO1_Escarpia_spicata", "KJ566960_CO1_Escarpia_sp._SP-2014", "U74063_CO1_Escarpia_laminata", "JN021267_CO1_Escarpia_sp._1_JMP-2012", 
#                                   "KP203873_CO1_Escarpia_sp._MNHN_IU-2013-77", "KP203874_CO1_Escarpia_sp._MNHN_IU-2013-78", "D50594_CO1_Paraescarpia_cf._echinospica", 
#                                   "D50595_CO1_Paraescarpia_cf._echinospica", "AB088670_CO1_cf._Alaysia_sp._A1", "AB088671_CO1_cf._Alaysia_sp._A2", 
#                                   "AB088672_CO1_cf._Alaysia_sp._A3", "FJ667536_CO1_Alaysia_sp._SBJ-2009")
CO1_ML.unconstrained.boottrees <- read.tree('~/jsb439@cornell.edu/amelia-bivalves/tubeworms/IQTREE/unconstrained_fullboot/CP.txt.boottrees')

#CO1_ML.unconstrained$tip.label==unlist(lapply(strsplit(swaps, split="_"), `[`, 1))
#this looks ok
#CO1_ML.unconstrained$tip.label<-swaps

CO1_ML.constrained<-read.tree('~/jsb439@cornell.edu/amelia-bivalves/tubeworms/IQTREE/constrained/constrained.treefile')
# swaps<-c("AB088673_CO1_cf._Alaysia_sp._A4", "AB073491_CO1_Arcovestia_ivanovi", "AY129124_CO1_Lamellibrachia_cf._luymesi", 
#                                 "AY129132_CO1_Lamellibrachia_cf._luymesi", "EU046616_CO1_Lamellibrachia_anaximandri", "HQ154525_CO1_Lamellibrachia_sp._Lam1", 
#                                 "HQ154526_CO1_Lamellibrachia_sp._Lam2", "HQ396891_CO1_Lamellibrachia_sp._DBUA_01146", "MH670807_CO1_Lamellibrachia_donwalshi", 
#                                 "KJ566961_CO1_Lamellibrachia_sp._2_SP-2014", "D38030_CO1_Lamellibrachia_satsuma", "U74078_CO1_Basibranchia_sp.__Mariana_Trough_", 
#                                 "U74055_CO1_Lamellibrachia_barhami", "D38029_CO1_Lamellibrachia_sp.", "LC064365_CO1_Lamellibrachia_sp._GK-2015", 
#                                 "D50592_CO1_Lamellibrachia_sp.", "U74061_CO1_Lamellibrachia_columna", "AB055209_CO1_Lamellibrachia_sp._L4", 
#                                 "AB088674_CO1_Lamellibrachia_sp._L6", "AB055210_CO1_Lamellibrachia_sp._L5", "AB088675_CO1_Lamellibrachia_sp._L7", 
#                                 "AB264601_CO1_Lamellibrachia_juni", "FM178480_CO1_Sclerolinum_contortum", "FJ347644_CO1_Sclerolinum_brattstromi", 
#                                 "U74068_CO1__Loihi_Seamount__perviate_pogonophoran", "FJ480347_CO1_Siboglinum_sp._1_SBJ-2009", 
#                                 "FJ480378_CO1_Siboglinum_sp._2_SBJ-2009", "FJ480394_CO1_Siboglinum_sp._3_SBJ-2009", "FJ480396_CO1_Siboglinum_sp._4_SBJ-2009", 
#                                 "FJ480358_CO1_Siboglinum_sp._5_SBJ-2009", "FJ480364_CO1_Siboglinum_sp._6_SBJ-2009", "FJ480360_CO1_Sibloglinidae_cf._Cyclobrachia_sp._SBJ-2009", 
#                                 "FJ480399_CO1_Siboglinum_poseidoni", "KJ789169_CO1_Siboglinum_ekmani", "FJ480389_CO1_Polybrachia_sp._2_SBJ-2009", 
#                                 "FJ480391_CO1_Polybrachia_sp._3_SBJ-2009", "FJ480393_CO1_Polybrachia_sp._1_SBJ-2009", "KJ789162_CO1_Galathealinum_brachiosum", 
#                                 "U74066_CO1_Galathealinum_brachiosum", "FJ480357_CO1_Bobmarleya_gadensis", "FJ480370_CO1_Spirobrachia_tripeira", 
#                                 "FJ483547_CO1_Spirobrachia_cf._grandis_SJ-2009", "KJ789171_CO1_Spirobrachia_sp._YL-2014", "FJ480376_CO1_Lamellisabella_denticulata", 
#                                 "KJ789170_CO1_Siboglinum_fiordicum", "FM178481_CO1_Oligobrachia_haakonmosbiensis", "MH619675_CO1_Oligobrachia_sp._1125-3", 
#                                 "FJ480388_CO1_Lamellisabella_sp._SBJ-2009", "AB259569_CO1_Osedax_japonicus", "DQ996629_CO1_Osedax_sp._3_SBJ-2006", 
#                                 "EU223340_CO1_Osedax_sp.__orange-collar_", "AY827562_CO1_Osedax_mucofloris", "FJ347615_CO1_Osedax_sp.__white-collar_", 
#                                 "FM998109_CO1_Osedax_sp._sagami-7", "JX280613_CO1_Osedax_sp.__MB16_", "AY586495_CO1_Osedax_frankpressi", 
#                                 "LC106303_CO1_Osedax_sp._1336_61_2", "LC381421_CO1_Osedax_braziliensis", "EU223307_CO1_Osedax_rubiplumus", 
#                                 "FM998082_CO1_Osedax_sp._sagami-4", "FM998081_CO1_Osedax_sp._sagami-3", "JF509949_CO1_Osedax_roseus", "JX280609_CO1_Osedax_sp.__MB17_", 
#                                 "EU223356_CO1_Osedax_sp.__nude-palp-A_", "EU236218_CO1_Osedax_sp.__nude-palp-B_", "KP119563_CO1_Osedax_sp._Nude-palp-G", 
#                                 "FJ347643_CO1_Osedax_sp.__nude-palp-F_", "KJ598040_CO1_Osedax_rogersi", "KF444422_CO1_Osedax_antarcticus", 
#                                 "KJ598038_CO1_Osedax_crouchi", "KJ598039_CO1_Osedax_nordenskjoeldi", "EU267675_CO1_Osedax_sp.__nude-palp-C_", 
#                                 "FM998107_CO1_Osedax_sp._sagami-6", "FJ347630_CO1_Osedax_sp.__nude-palp-D_", "FM998110_CO1_Osedax_sp._sagami-8", 
#                                 "FJ347632_CO1_Osedax_sp.__nude-palp-E_", "FJ347639_CO1_Osedax_sp.__green-palp_", "FM998083_CO1_Osedax_sp._sagami-5", 
#                                 "KT860548_CO1_Osedax_sp.__Mediterranea_", "KP119564_CO1_Osedax_priapus", "FJ347616_CO1_Osedax_sp.__yellow-patch_", 
#                                 "KT860545_CO1_Osedax_deceptionensis", "DQ996622_CO1_Osedax_sp._1_SBJ-2006", "JF317197_CO1_Hesiospina_aurantiaca", 
#                                 "MH202758_CO1_Amphisamytha_sp._n._1_YZ-2018", "MH202759_CO1_Amphisamytha_sp._n._2_YZ-2018", "KP144926_CO1_Nephtys_hombergii", 
#                                 "AY428839_CO1_Owenia_fusiformis", "MH202756_CO1_Hesiolyra_cf._bergi_YZ-2018", "AB242857_CO1_Oasisia_fujikurai", 
#                                 "FJ667533_CO1_Oasisia_sp._SBJ-2009", "U74069_CO1_Oasisia_alvinae", "FJ667534_CO1_Oasisia_sp._SBJ-2009", 
#                                 "FJ667535_CO1_Oasisia_sp._SBJ-2009", "U74056_CO1_Ridgeia_sp.__Southern_Explorer_Ridge_", "U74073_CO1_Ridgeia_piscesae", 
#                                 "U74075_CO1_Tevnia_jerichonana", "U74074_CO1_Riftia_pachyptila", "AF317287_CO1_Seepiophila_jonesi", 
#                                 "AY129134_CO1_Escarpia_sp._GB425-GoM", "AY326303_CO1_Escarpia_southwardae", "KJ566960_CO1_Escarpia_sp._SP-2014", 
#                                 "U74065_CO1_Escarpia_spicata", "U74063_CO1_Escarpia_laminata", "JN021267_CO1_Escarpia_sp._1_JMP-2012", 
#                                 "KP203873_CO1_Escarpia_sp._MNHN_IU-2013-77", "KP203874_CO1_Escarpia_sp._MNHN_IU-2013-78", "D50594_CO1_Paraescarpia_cf._echinospica", 
#                                 "D50595_CO1_Paraescarpia_cf._echinospica", "AB088670_CO1_cf._Alaysia_sp._A1", "AB088671_CO1_cf._Alaysia_sp._A2", 
#                                 "AB088672_CO1_cf._Alaysia_sp._A3", "FJ667536_CO1_Alaysia_sp._SBJ-2009")

CO1_ML.constrained.boottrees <- read.tree('~/jsb439@cornell.edu/amelia-bivalves/tubeworms/IQTREE/constrained_fullboot/constrained.boottrees')


# CO1_ML.constrained$tip.label==unlist(lapply(strsplit(swaps, split="_"), `[`, 1))
# #this looks ok
# CO1_ML.constrained$tip.label<-swaps

}

# #testing
# assoc<-(cbind(CO1_ML.constrained$tip.label,CO1_ML.constrained$tip.label))
# obj<- cophylo(CO1_ML.unconstrained, CO1_ML.constrained, assoc=assoc, rotate=T )
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)

#data checks
{
#checking that the tip names match the correct rows in the data table
sort(c(ingroup_pruned$TaxID, outgroup$TaxID))==sort(CO1_ML.unconstrained$tip.label)
#checking that the tip names match the correct rows in the data table
sort(c(ingroup_pruned$TaxID, outgroup$TaxID))==sort(CO1_ML.constrained$tip.label)

#CO1_ML.constrained$tip.label==CO1_ML.unconstrained$tip.label
}

#processing
{
#combined data for ingroup and outgorup
data<-rbind(ingroup_pruned,outgroup)
#convert data to data frame
data<-as.data.frame(data)
rownames(data)<-data$TaxID
}

#length(data$order)

# sort(rownames(data))==sort(CO1_ML.unconstrained$tip.label)
# sort(rownames(data))==sort(CO1_ML.constrained$tip.label)

#rerooting
{
root.unconstrained<-findMRCA(CO1_ML.unconstrained, tips=outgroup$TaxID) #untrimmed
root.constrained<-findMRCA(CO1_ML.constrained, tips=outgroup$TaxID) #untrimmed

#checking
plot(CO1_ML.unconstrained, cex=0.001)
nodelabels(node=root.unconstrained)

plot(CO1_ML.constrained, cex=0.001)
nodelabels(node=root.constrained)

#reroot on root
tree.unconstrained <- reroot(CO1_ML.unconstrained, node.number=root.unconstrained)
tree.constrained <- reroot(CO1_ML.constrained, node.number=root.constrained)

plot(tree.unconstrained)
plot(tree.constrained)

}

# par(mfrow=c(1,2))
# plot(ladderize(tree.unconstrained), cex=0.01)
# plot(ladderize(tree.constrained), cex=0.01)

# 
# #checking
# assoc<-(cbind(tree.unconstrained$tip.label,tree.unconstrained$tip.label))
# obj<- cophylo(tree.unconstrained, tree.constrained, assoc=assoc, rotate=T )
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)

#prune outgroups
{
#cut the outgroup tips
prunetree.constrained <- drop.tip(tree.constrained, tip=outgroup$TaxID)
prunetree.unconstrained <- drop.tip(tree.unconstrained, tip=outgroup$TaxID)
}

# 
# #checking
# assoc<-(cbind(prunetree.constrained$tip.label,prunetree.constrained$tip.label))
# obj<- cophylo(prunetree.constrained, prunetree.unconstrained, assoc=assoc, rotate=T )
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)

# #calculate root to tip lengths #commenting out for now
{
#root to tip length from the pruned tree as r2t
r2t.prunetree.constrained<-as.data.frame(diag(vcv.phylo(prunetree.constrained)))
#ensure they are the same order
r2t.prunetree.unconstrained <- (diag(vcv.phylo(prunetree.unconstrained)))
r2t.prunetree.unconstrained <- as.data.frame(r2t.prunetree.unconstrained[rownames(r2t.prunetree.constrained)])
}

#check constrained vs unconstrained bls
#plot(log(r2t.prunetree.constrained$`diag(vcv.phylo(prunetree.constrained))`)~ log(r2t.prunetree.unconstrained$`r2t.prunetree.unconstrained[rownames(r2t.prunetree.constrained)]`))
#constrained v unconstrained is mostly similar, some differences due to differences in root position

#data processing
{
#prune out outgroups
data<-data[!data$is.outgroup==1,]

#reorder the data frame to match r2t
#this also prunes out the outgroups from the dataset
data.constrained<-data[rownames(r2t.prunetree.constrained),]
data.unconstrained<-data[rownames(r2t.prunetree.unconstrained),]

#rownames(r2t.prunetree.constrained) == rownames(r2t.prunetree.unconstrained)
#checking names
#data.frame(sort(rownames(data.constrained)), sort(rownames(r2t.prunetree.constrained)), c(sort(rownames(data.constrained))== sort(rownames(r2t.prunetree.constrained))))
#data.frame(sort(rownames(data.unconstrained)), sort(rownames(r2t.prunetree.unconstrained)), c(sort(rownames(data.unconstrained))== sort(rownames(r2t.prunetree.unconstrained))))


#prune out rows with multiple habitat types
#get rid of Seep, OF
remove<-rownames(data.constrained[data.constrained$Habitat_Mod=='Seep, OF',]) #not being stored
data.constrained <- data.constrained[!data.constrained$Habitat_Mod=='Seep, OF',]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=remove)

remove<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=='Seep, OF',]) #not being stored
data.unconstrained <- data.unconstrained[!data.unconstrained$Habitat_Mod=='Seep, OF',]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=remove)


#get rid of vent, seep
remove<-rownames(data.constrained[data.constrained$Habitat_Mod=='Vent, Seep',])
data.constrained <- data.constrained[! data.constrained$Habitat_Mod=='Vent, Seep',]
prunetree.constrained<- drop.tip(prunetree.constrained, tip=remove)

remove<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=='Vent, Seep',])
data.unconstrained <- data.unconstrained[! data.unconstrained$Habitat_Mod=='Vent, Seep',]
prunetree.unconstrained<- drop.tip(prunetree.unconstrained, tip=remove)


#get rid of unknown 
remove<-rownames(data.constrained[data.constrained$Habitat_Mod=='?',])
data.constrained <- data.constrained[!data.constrained$Habitat_Mod=='?',]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=remove)

remove<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=='?',])
data.unconstrained <- data.unconstrained[!data.unconstrained$Habitat_Mod=='?',]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=remove)


#get rid of Vent, Seep, OF
VSO<-rownames(data.constrained[data.constrained$Habitat_Mod=='Vent, Seep, OF',])
data.constrained <- data.constrained[!data.constrained$Habitat_Mod=='Vent, Seep, OF',]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=VSO)

VSO<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=='Vent, Seep, OF',])
data.unconstrained <- data.unconstrained[!data.unconstrained$Habitat_Mod=='Vent, Seep, OF',]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=VSO)


#get rid of Vent, Seep, Shallow
VSS<-rownames(data.constrained[data.constrained$Habitat_Mod=='Vent, Seep, Shallow',])
data.constrained <- data.constrained[!data.constrained$Habitat_Mod=='Vent, Seep, Shallow',]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=VSS)

VSS<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=='Vent, Seep, Shallow',])
data.unconstrained <- data.unconstrained[!data.unconstrained$Habitat_Mod=='Vent, Seep, Shallow',]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=VSS)


#get rid of Ambient
AMB<-rownames(data.constrained[data.constrained$Habitat_Mod=='Ambient',])
data.constrained <- data.constrained[!data.constrained$Habitat_Mod=='Ambient',]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=AMB)

AMB<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=='Ambient',])
data.unconstrained <- data.unconstrained[!data.unconstrained$Habitat_Mod=='Ambient',]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=AMB)


#get rid of Vent, Seep, Ambient
VSA<-rownames(data.constrained[data.constrained$Habitat_Mod=="Vent, Seep, Ambient",])
data.constrained <- data.constrained[!data.constrained$Habitat_Mod=="Vent, Seep, Ambient",]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=VSA)

VSA<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=="Vent, Seep, Ambient",])
data.unconstrained <- data.unconstrained[!data.unconstrained$Habitat_Mod=="Vent, Seep, Ambient",]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=VSA)


#get rid of Seep, Ambient
SA<-rownames(data.constrained[data.constrained$Habitat_Mod=="Seep, Ambient",])
data.constrained <- data.constrained[!data.constrained$Habitat_Mod=="Seep, Ambient",]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=SA)

SA<-rownames(data.unconstrained[data.unconstrained$Habitat_Mod=="Seep, Ambient",])
data.unconstrained <- data.unconstrained[!data.unconstrained$Habitat_Mod=="Seep, Ambient",]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=SA)



#remove frenulates
#get rid of Seep, Ambient
fren<-rownames(data.constrained[data.constrained$frenulate=="1",])
data.constrained <- data.constrained[!data.constrained$frenulate=="1",]
prunetree.constrained <- drop.tip(prunetree.constrained, tip=fren)

fren<-rownames(data.unconstrained[data.unconstrained$frenulate=="1",])
data.unconstrained <- data.unconstrained[!data.unconstrained$frenulate=="1",]
prunetree.unconstrained <- drop.tip(prunetree.unconstrained, tip=fren)


####
#get rid of everything that is not Vestimentifera
#NVES<-rownames(data[data$Lineage=='Vestimentifera',])
#data <- data[prunetree$tip.label %in% NVES,]
#prunetree<- drop.tip(prunetree, tip=prunetree$tip.label[!prunetree$tip.label %in% NVES])
####

#data.unconstrained$Lineage
}

# 
# plot(prunetree.unconstrained, cex=0.1, no.margin=T)
# plot(prunetree.constrained, cex=0.1, no.margin=T)


# #checking
# assoc<-(cbind(prunetree.constrained$tip.label,prunetree.constrained$tip.label))
# obj<- cophylo(prunetree.constrained, prunetree.unconstrained, assoc=assoc, rotate=T)
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)

#store data
{
#root to tip length from the pruned tree as r2t
r2t.constrained<-as.data.frame(diag(vcv.phylo(prunetree.constrained)))
#ensure they are the same order
r2t.unconstrained <- (diag(vcv.phylo(prunetree.unconstrained)))
r2t.unconstrained <- as.data.frame(r2t.unconstrained[rownames(r2t.constrained)])

#plot(log(r2t.unconstrained$`r2t.unconstrained[rownames(r2t.constrained)]`)~ log(r2t.constrained$`diag(vcv.phylo(prunetree.constrained))`))
#constrained v unconstrained is quite a large difference
}

# #plot(prunetree, cex=0.4, no.margin=T)
# length(data.constrained[,1]) #checks how many rows in the data
# prunetree.constrained #check how many tips and nodes
# 
# length(data.unconstrained[,1]) #checks how many rows in the data
# prunetree.unconstrained #check how many tips and nodes

#store data
{
#store habitat data
x.constrained <- data.constrained$Habitat_Mod
names(x.constrained) <- rownames(data.constrained)

x.unconstrained <- data.unconstrained$Habitat_Mod
names(x.unconstrained) <- rownames(data.unconstrained)

#store r2t data
y.constrained <- r2t.constrained[,1]
names(y.constrained) <- rownames(r2t.constrained)

y.unconstrained <- r2t.unconstrained[,1]
names(y.unconstrained) <- rownames(r2t.unconstrained)

#store size data
z.unconstrained<-data.unconstrained$tube_mm
z.constrained<-data.constrained$tube_mm

names(z.unconstrained)<-rownames(r2t.unconstrained)
names(z.constrained)<-rownames(r2t.constrained)

##
#store diaphrahm data
dia.unconstrained<-data.unconstrained$`diaphragm_mm(vest or fore); "trunk" in osedax`
dia.constrained<-data.constrained$`diaphragm_mm(vest or fore); "trunk" in osedax`

names(dia.unconstrained)<-rownames(data.unconstrained)
names(dia.constrained)<-rownames(data.constrained)

###store size data
palp.unconstrained<-data.unconstrained$obt_palp_tent_12.28.22
palp.constrained<-data.constrained$obt_palp_tent_12.28.22

names(palp.unconstrained)<-rownames(data.unconstrained)
names(palp.constrained)<-rownames(data.unconstrained)


###


spread.unconstrained<-data.unconstrained$rate_cat
spread.constrained<-data.constrained$rate_cat

names(spread.unconstrained)<-rownames(data.unconstrained)
names(spread.constrained)<-rownames(data.constrained)


#checking names again
names(x.unconstrained) == names(y.unconstrained)
names(x.constrained) == names(y.constrained)

##### 
#tmp<-data.frame(r2t=y, habitat=x, size=z, spread=spread)
#write.csv(tmp, file='r2t_habitat_size_spread.txt', row.names=T)
#####
}


#generate congruified tree
{
YuanningLe<-read.tree("~/jsb439@cornell.edu/amelia-bivalves/tubeworms/constraint_2_19_22.tre")
is.ultrametric(YuanningLe)

#YuanningLe<-drop.tip(YuanningLe)
#drop tips that have been pruned
YuanningLe <- drop.tip(YuanningLe, tip=c("FJ347644_CO1_Sclerolinum_brattstromi", "KJ789169_CO1_Siboglinum_ekmani", "KJ789170_CO1_Siboglinum_fiordicum", "U74065_CO1_Escarpia_spicata", 
                                         "HQ023477_CO1_Cirratulus_aff_cirratus_CMC02", "KY472732_CO1_Sabella_spallanzanii", "MN256592_CO1_Sternaspis_scutata", fren))


tax<-data.frame(row.names = YuanningLe$tip.label, taxonomy=YuanningLe$tip.label)

#YuanningLe$tip.label %in% prunetree.constrained$tip.label
#drop<-prunetree.constrained$tip.label[!prunetree.constrained$tip.label %in% YuanningLe$tip.label]

congruify.constrained<-congruify.phylo(YuanningLe, prunetree.constrained, taxonomy=tax, scale='treePL')
congruify.unconstrained<-congruify.phylo(YuanningLe, prunetree.unconstrained, taxonomy=tax, scale='treePL')

# par(mfrow=c(1,3))
# plot(congruify.constrained$phy, cex=0.5)
# axisPhylo()
# plot(prunetree.constrained, cex=0.5)
# axisPhylo()
# plot(YuanningLe, cex=0.8)
# axisPhylo()
# 
# #testing 
# assoc<-(cbind(YuanningLe$tip.label,YuanningLe$tip.label))
# obj<- cophylo(prunetree.unconstrained, YuanningLe, assoc=assoc, rotate=T )
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)
# 
# #testing 
# assoc<-(cbind(YuanningLe$tip.label,YuanningLe$tip.label))
# obj<- cophylo(prunetree.constrained, YuanningLe, assoc=assoc, rotate=T )
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)
# 
# assoc<-(cbind(prunetree.constrained$tip.label,prunetree.constrained$tip.label))
# obj<- cophylo(prunetree.constrained, prunetree.unconstrained, assoc=assoc, rotate=T )
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)
# 
# assoc<-(cbind(sort(drop.tip(prunetree.constrained, drop)$tip.label),sort(drop.tip(prunetree.unconstrained, drop)$tip.label)))
# obj<- cophylo(drop.tip(prunetree.constrained, drop), drop.tip(prunetree.unconstrained, drop), assoc=assoc, rotate=T )
# plot(obj,link.type="curved",link.lwd=3,link.lty="solid",link.col=make.transparent("blue",0.25),fsize=0.3)
# 
# #this looks OK, and it looks like the constraint worked (but it may not be very effective givein the limited sampling)
# 
# #phytools r package anova
# #uses the raxml tree
# #nsim is the number of simulations, default is 1000
# #increasing the number seems to give more power
# 
# #generating time tree for anova
# #prunetree.time<-chronos(prunetree, model='relaxed')
# #prunetree.time<-ladderize(prunetree.time, right=F)
# #plot(prunetree.time, cex=0.5)
# #axisPhylo()
}

#testFitchBeintema(prunetree.constrained)
#testFitchBeintema(prunetree.unconstrained)
#testFitchBeintemaNLME(prunetree.unconstrained)


#checking for node density artifact
path_lengths <- diag(vcv(prunetree.constrained))
node_paths <- nodepath(prunetree.constrained)
node_counts <- sapply(node_paths, length) - 1

node_data <- data.frame(
  NodeCounts = node_counts,  # x variable (speciation events from root)
  TotalPathLength = path_lengths  # y variable (genetic distance to root)
)

plot(node_data$TotalPathLength ~ node_data$NodeCounts)
abline(lm(node_data$TotalPathLength ~ node_data$NodeCounts))
summary(lm(node_data$TotalPathLength ~ node_data$NodeCounts))

# > summary(lm(node_data$TotalPathLength ~ node_data$NodeCounts))
# 
# Call:
#   lm(formula = node_data$TotalPathLength ~ node_data$NodeCounts)
# 
# Residuals:
#   Min       1Q   Median       3Q      Max 
# -0.24030 -0.09607  0.01565  0.08947  0.24144 
# 
# Coefficients:
#   Estimate Std. Error t value Pr(>|t|)    
# (Intercept)          0.621863   0.063597   9.778 7.56e-15 ***
#   node_data$NodeCounts 0.017894   0.008603   2.080   0.0411 *  
#   ---
#   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
# 
# Residual standard error: 0.1188 on 72 degrees of freedom
# Multiple R-squared:  0.05668,	Adjusted R-squared:  0.04358 
# F-statistic: 4.326 on 1 and 72 DF,  p-value: 0.04108


#####FOR RUNNING TRAITRATE#####

##setting up three comparisons ### 
#vent vs OF, remove seep, comparison 1
#vent vs seep, remove OF, comparison 2
#seep vs OF, remove vent, comparison 3
#require(ips)
{
  #set up the genetic datasets
  fas<-read.FASTA("/Users/cotinga/jsb439@cornell.edu/amelia-bivalves/tubeworms/IQTREE/mafft_aligned_2_19_22_2-sequencher.fas")
  
  
  #set up comparison 1
  
  #prune out seep
  remove1<-rownames(data[data$Habitat_Mod=='Seep',])
  #remove seep tips from tree
  traitRate1 <- drop.tip(congruify.constrained$phy, tip=remove1)
  #write.tree(traitRate1, file="traitRate1.tre")
  
  #create trait dataset
  trim1 <- data.constrained[!data.constrained$Habitat_Mod=='Seep',]
  trait1 <- as.numeric(as.factor(trim1$Habitat_Mod))-1
  trait1 <- data.frame(trait= setNames(trait1, 
                                       rownames(trim1))[traitRate1$tip.label])
  #ips:::write.fas(trait1,file="trait1.fas")
  
  #extract the seqs
  traitRate1.seq<-fas[traitRate1$tip.label]
  #ips:::write.fas(traitRate1.seq, file="trait1.seq.fas")
  
  
  
  #set up comparison 2
  
  #prune out seep
  remove2<-rownames(data.constrained[data.constrained$Habitat_Mod=='OF',])
  #remove seep tips from tree
  traitRate2 <- drop.tip(congruify.constrained$phy, tip=remove2)
  #write.tree(traitRate2, file="traitRate2.tre")
  
  #create trait dataset
  trim2 <- data.constrained[!data.constrained$Habitat_Mod=='OF',]
  trait2 <- as.numeric(as.factor(trim2$Habitat_Mod))-1
  trait2<-data.frame(trait= setNames(trait2, 
                                     rownames(trim2))[traitRate2$tip.label])
  #ips:::write.fas(trait2,file="trait2.fas")
  
  #extract the seqs
  traitRate2.seq<-fas[traitRate2$tip.label]
  #ips:::write.fas(traitRate2.seq, file="trait2.seq.fas")
  
  
  #set up comparison 3
  
  #prune out vent
  remove3<-rownames(data.constrained[data.constrained$Habitat_Mod=='Vent',])
  #remove seep tips from tree
  traitRate3 <- drop.tip(congruify.constrained$phy, tip=remove3)
  #write.tree(traitRate3, file="traitRate3.tre")
  
  
  #create trait dataset
  trim3 <- data.constrained[!data.constrained$Habitat_Mod=='Vent',]
  trait3 <- as.numeric(as.factor(trim3$Habitat_Mod))-1
  trait3<-data.frame(trait= setNames(trait3, 
                                     rownames(trim3))[traitRate3$tip.label])
  #ips:::write.fas(trait3,file="trait3.fas")
  
  #extract the seqs
  traitRate3.seq<-fas[traitRate3$tip.label]
  #ips:::write.fas(traitRate3.seq, file="trait3.seq.fas")
  
  
  
  
  # 
  # traitRate(phy=traitRate1, seq=as.matrix(traitRate1.seq), x=trait1, 
  #           exec = "/Applications/Phylogenetics/TraitRateProp_source_20161221/programs/traitRateProp")
  # 
  # traitRate(phy=traitRate2, seq=as.matrix(traitRate2.seq), x=trait2, 
  #           exec = "/Applications/Phylogenetics/TraitRateProp_source_20161221/programs/traitRateProp")
  # 
}

#processing traitrate output
{
  #read in traitRate results
  traitRateRes1<-read.csv(file="/Users/cotinga/jsb439@cornell.edu/amelia-bivalves/tubeworms/traitRate1_nofren/TraitRateProp_per_site_report.csv")
  traitRateRes1.bayes<-traitRateRes1$bayes_factor
  traitRateRes1.bayes[traitRateRes1.bayes < 3] <- -2
  traitRateRes1.bayes[traitRateRes1.bayes > 20] <- 20
  
  #append 42 blanks to get it in the right frame
  #traitRateRes1.bayes<- c(rep(-10,42), traitRateRes1.bayes)
  
  
  traitRateRes2<-read.csv(file="/Users/cotinga/jsb439@cornell.edu/amelia-bivalves/tubeworms/traitRate2_nofren/TraitRateProp_per_site_report.csv")
  traitRateRes2.bayes <- traitRateRes2$bayes_factor
  traitRateRes2.bayes[traitRateRes2.bayes < 3] <- -2
  traitRateRes2.bayes[traitRateRes2.bayes > 20 ] <- 20
  #append 42 blanks to get it in the right frame
  #traitRateRes2.bayes<- c(rep(-10,42), traitRateRes2.bayes)
  
  
  traitRateRes3<-read.csv(file="/Users/cotinga/jsb439@cornell.edu/amelia-bivalves/tubeworms/traitRate3_nofren/TraitRateProp_per_site_report.csv")
  traitRateRes3.bayes <- traitRateRes3$bayes_factor
  traitRateRes3.bayes[traitRateRes3.bayes < 3] <- -2
  traitRateRes3.bayes[traitRateRes3.bayes > 20] <- 20
  
  #append 42 blanks to get it in the right frame
  #traitRateRes3.bayes<- c(rep(-10,42), traitRateRes3.bayes)
  
}

#set up simmap visualizations to show the trait distributions
{
  traitRate1.simmap<-make.simmap(traitRate1,setNames(trait1$trait, rownames(trait1)),
                                 model="ARD", nsim=100)
  traitRate1.simmap.obj<-densityMap(traitRate1.simmap,states=levels(setNames(trait1$trait, rownames(trait1)))[2:1],plot=FALSE)
  
  traitRate2.simmap<-make.simmap(traitRate2,setNames(trait2$trait, rownames(trait2)),
                                 model="ARD", nsim=100)
  traitRate2.simmap.obj<-densityMap(traitRate2.simmap,states=levels(setNames(trait2$trait, rownames(trait2)))[2:1],plot=FALSE)
  
  traitRate3.simmap<-make.simmap(traitRate3,setNames(trait3$trait, rownames(trait3)),
                                 model="ARD", nsim=100)
  traitRate3.simmap.obj<-densityMap(traitRate3.simmap,states=levels(setNames(trait3$trait, rownames(trait3)))[2:1],plot=FALSE)
}


require('WGCNA')
#code to generate traitRate plots

pdf(file="traitRate_tubewormsCO1_no_frenulates.pdf", height=4, width = 8.5)
{
  
  #set up the plotting region for traitrate
  {
    layout.matrix <- matrix(c(1,3,5,2,4,6), nrow = 3, ncol = 2)
    
    layout(mat = layout.matrix,
           heights = c(1, 1, 1), # Heights
           widths = c(0.35, 2.5, 2.5)) # Widths of the two columns
  }
  
  #par(mfrow=c(3,1))
  #par(mar=c(7.1, 4.1, 7.1, 2.1), xpd=F)
  
  #labels reflect CO1 position in
  #NC_056377.1:14737-16287_Mytella_strigata
  #plot 1
  {
    plot(traitRate1.simmap.obj, legend=F, ftype="off", lwd=1,
         mar=c(1.1, 1, 2, 0.5))
    
    par(mar=c(1, 4.1, 2, 2.1), xpd=F)
    
    plot(traitRateRes1.bayes, 
         ylab="BF Vent vs OF", ylim=c(2,20), 
         bty='n', xlim=c(0,1200),
         pch=19,
         col=rep(c("red", "red", "black"), 508),
         xaxt="n", yaxt='n',
         xlab="COX1 position", main="COX1 site position")
    axis(1, at = c(1, 200, 400, 600, 800, 1000, 1200)-1, 
         labels = rep("", 7))#,c(1, 100, 200, 300, 400, 500, 600, 700))
    axis(2, at = c(3,10,15,20), las=2, cex.axis=0.8)
    #lines(traitRateRes1.bayes, col = "black")
    labelPoints(
      x=1:length(traitRateRes1.bayes), y=traitRateRes1.bayes, 
      labels=1:length(traitRateRes1.bayes), 
      cex = 0.6, offs = 0.05, xpd = F, 
      jiggle = 0, protectEdges = F, 
      doPlot = TRUE, col="black")
  }
  
  #plot 2
  {
    plot(traitRate2.simmap.obj, legend=F, ftype="off", lwd=1,
         mar=c(1.1, 1, 2, 0.5))
    
    par(mar=c(1, 4.1, 2, 2.1), xpd=F)
    
    plot(traitRateRes2.bayes, 
         ylab="BF Vent vs Seep", ylim=c(2,20), 
         bty='n', xlim=c(0,1200),
         pch=19,
         col=rep(c("red", "red", "black"), 508),
         xaxt="n", yaxt='n',
         xlab="COX1 position")
    axis(1, at = c(1, 200, 400, 600, 800, 1000, 1200)-1, 
         labels = rep("", 7))
    axis(2, at = c(3,10,15,20), las=2, cex.axis=0.8)
    labelPoints(
      x=1:length(traitRateRes2.bayes), y=traitRateRes2.bayes, 
      labels=1:length(traitRateRes2.bayes), 
      cex = 0.6, offs = 0.05, xpd = F, 
      jiggle = 0, protectEdges = F, 
      doPlot = TRUE)
    #lines(traitRateRes2.bayes, col = "black")
    
  }
  
  #plot 3
  {
    plot(traitRate3.simmap.obj, legend=F, ftype="off", lwd=1,
         mar=c(2, 1, 1.5, 0.5))
    
    par(mar=c(2, 4.1, 1, 2.1), xpd=F)    
    plot(traitRateRes3.bayes, 
         ylab="BF Seep vs OF", ylim=c(2,23), 
         bty='n', xlim=c(0,1200),
         pch=19,
         col=rep(c("red", "red", "black"), 508),
         xaxt="n", yaxt='n',
         xlab="COX1 position")
    axis(1, at = c(1, 200, 400, 600, 800, 1000, 1200)-1, 
         labels = c(1, 200, 400, 600, 800, 1000, 1200))
    axis(2, at = c(3,10,15,20), las=2, cex.axis=0.8)
    labelPoints(
      x=1:length(traitRateRes3.bayes), y=traitRateRes3.bayes, 
      labels=1:length(traitRateRes3.bayes), 
      cex = 0.6, offs = 0.05, xpd = F, 
      jiggle = 0, protectEdges = F, 
      doPlot = TRUE)
    #lines(traitRateRes3.bayes, col = "black")
    
  }
  
}
dev.off()


#what % of significant sites are first or second codon positions.
mean(c(
  5/sum(traitRateRes1.bayes >= 1),
  3/sum(traitRateRes2.bayes >= 1),
  2/sum(traitRateRes3.bayes >= 1)
))

sd(c(
  5/sum(traitRateRes1.bayes >= 1),
  3/sum(traitRateRes2.bayes >= 1),
  2/sum(traitRateRes3.bayes >= 1)
))

#####END FOR RUNNING TRAITRATE#####

boxplot((log(y.unconstrained))~x.constrained, varwidth=T)
boxplot((log(y.constrained))~x.constrained, varwidth=T)

#store data
{
  prunetree.constrained.time<-congruify.constrained$phy
  prunetree.unconstrained.time<-congruify.unconstrained$phy
  
  #reorder
  x.unconstrained<-x.unconstrained[prunetree.constrained.time$tip.label]
  x.constrained<-x.constrained[prunetree.constrained.time$tip.label]
  y.unconstrained<-y.unconstrained[prunetree.constrained.time$tip.label]
  y.constrained<-y.constrained[prunetree.constrained.time$tip.label]
  z.constrained<-z.constrained[prunetree.constrained.time$tip.label]
  z.unconstrained<-z.unconstrained[prunetree.constrained.time$tip.label]
}

#uses the 'time calibrated tree' 
#run phylogenetic anovas
{
  phyl.aov.constrained <- phylANOVA(prunetree.constrained.time, x = x.unconstrained, y = log(y.constrained), nsim=10000, posthoc=T, p.adj = 'BH')
  
  # ANOVA table: Phylogenetic ANOVA
  # 
  # Response: y
  # Sum Sq  Mean Sq   F value Pr(>F)
  # x        1.032396 0.516198 36.534317 0.3148
  # Residual 1.003168 0.014129                 
  # 
  # P-value based on simulation.
  # ---------
  #   
  #   Pairwise posthoc test using method = "BH"
  # 
  # Pairwise t-values:
  #   OF     Seep      Vent
  # OF    0.000000 7.616377  5.818119
  # Seep -7.616377 0.000000 -1.996350
  # Vent -5.818119 1.996350  0.000000
  # 
  # Pairwise corrected P-values:
  #   OF    Seep   Vent
  # OF   1.00000 0.33765 0.4111
  # Seep 0.33765 1.00000 0.3261
  # Vent 0.41110 0.32610 1.0000
  # ---------
  
  #phyl.aov.constrained <- phylANOVA(prunetree.constrained.time, x = x.unconstrained, y = log(y.constrained), nsim=10000, posthoc=T, p.adj = 'none')
  
  
  # ANOVA table: Phylogenetic ANOVA
  # 
  # Response: y
  # Sum Sq  Mean Sq   F value Pr(>F)
  # x        1.032396 0.516198 36.534317 0.3144
  # Residual 1.003168 0.014129                 
  # 
  # P-value based on simulation.
  # ---------
  #   
  #   Pairwise posthoc test using method = "none"
  # 
  # Pairwise t-values:
  #   OF     Seep      Vent
  # OF    0.000000 7.616377  5.818119
  # Seep -7.616377 0.000000 -1.996350
  # Vent -5.818119 1.996350  0.000000
  # 
  # Pairwise corrected P-values:
  #   OF   Seep   Vent
  # OF   1.0000 0.2271 0.4153
  # Seep 0.2271 1.0000 0.1045
  # Vent 0.4153 0.1045 1.0000
  # ---------
  #   

  phyl.aov.unconstrained <- phylANOVA(prunetree.constrained.time, x = x.unconstrained, y = log(y.unconstrained), nsim=1000, posthoc=T, p.adj = 'BH')
  
  # ANOVA table: Phylogenetic ANOVA
  # 
  # Response: y
  # Sum Sq  Mean Sq    F value Pr(>F)
  # x        3.504810 1.752405 126.600607  0.023
  # Residual 0.982782 0.013842                  
  # 
  # P-value based on simulation.
  # ---------
  #   
  #   Pairwise posthoc test using method = "BH"
  # 
  # Pairwise t-values:
  #   OF      Seep      Vent
  # OF     0.0000 12.851003 12.544301
  # Seep -12.8510  0.000000 -1.212154
  # Vent -12.5443  1.212154  0.000000
  # 
  # Pairwise corrected P-values:
  #   OF  Seep   Vent
  # OF   1.0000 0.036 0.0495
  # Seep 0.0360 1.000 0.3280
  # Vent 0.0495 0.328 1.0000
  # ---------
  
  
  #phyl.aov.unconstrained <- phylANOVA(prunetree.constrained.time, x = x.unconstrained, y = log(y.unconstrained), nsim=10000, posthoc=T, p.adj = 'none')
  
  # ANOVA table: Phylogenetic ANOVA
  # 
  # Response: y
  # Sum Sq  Mean Sq    F value Pr(>F)
  # x        3.504810 1.752405 126.600607 0.0279
  # Residual 0.982782 0.013842                  
  # 
  # P-value based on simulation.
  # ---------
  #   
  #   Pairwise posthoc test using method = "none"
  # 
  # Pairwise t-values:
  #   OF      Seep      Vent
  # OF     0.0000 12.851003 12.544301
  # Seep -12.8510  0.000000 -1.212154
  # Vent -12.5443  1.212154  0.000000
  # 
  # Pairwise corrected P-values:
  #   OF   Seep   Vent
  # OF   1.0000 0.0167 0.0429
  # Seep 0.0167 1.0000 0.3329
  # Vent 0.0429 0.3329 1.0000
  # ---------
  
  #phyl.aov.unconstrained.alt <- phylANOVA(prunetree.constrained.time, x = x.unconstrained, y = log(y.unconstrained), nsim=10000, posthoc=T)
  
  phyl.aov.unconstrained.alt <- phylANOVA(prunetree.unconstrained.time, x = x.unconstrained, y = log(y.unconstrained), nsim=10000, posthoc=T)
  
  # ANOVA table: Phylogenetic ANOVA
  # 
  # Response: y
  # Sum Sq  Mean Sq    F value Pr(>F)
  # x        3.504810 1.752405 126.600607 0.0216
  # Residual 0.982782 0.013842                  
  # 
  # P-value based on simulation.
  # ---------
  #   
  #   Pairwise posthoc test using method = "holm"
  # 
  # Pairwise t-values:
  #   OF      Seep      Vent
  # OF     0.0000 12.851003 12.544301
  # Seep -12.8510  0.000000 -1.212154
  # Vent -12.5443  1.212154  0.000000
  # 
  # Pairwise corrected P-values:
  #   OF   Seep   Vent
  # OF   1.0000 0.0384 0.0714
  # Seep 0.0384 1.0000 0.4475
  # Vent 0.0714 0.4475 1.0000
  # ---------
  
  
  phyl.aov.constrained
  phyl.aov.unconstrained
  phyl.aov.unconstrained.alt ## this generates some interesting results
  #phyl.aov.unconstrained.alt2
  
}

#phylogenetic anova from geiger r package
#run overall anova
{
  x.unconstrained.factor<-as.factor(x.unconstrained) #this functions needs x (the groups) to be a factor, 
  x.constrained.factor<-as.factor(x.constrained) #this functions needs x (the groups) to be a factor, 
  
  #using 'time calibrated tree'
  phyl.aov.geiger<-aov.phylo(log(y.constrained)~ x.constrained.factor, prunetree.constrained.time, nsim=10000)
  summary(phyl.aov.geiger)
  attr(phyl.aov.geiger, 'summary')
  
  # Analysis of Variance Table
  # 
  # Response: dat
  # Df  Sum-Sq  Mean-Sq F-value     Pr(>F) Pr(>F) given phy
  # group      2 0.55782 0.278909  38.188 5.5025e-12           0.3044
  # Residuals 71 0.51856 0.007304     
  # 
  
  
  #using 'time calibrated tree'
  phyl.aov.geiger<-aov.phylo(log(y.unconstrained)~ x.constrained.factor, prunetree.constrained.time, nsim=10000)
  summary(phyl.aov.geiger)
  attr(phyl.aov.geiger, 'summary')
  
  # Analysis of Variance Table
  # 
  # Response: dat
  # Df  Sum-Sq Mean-Sq F-value     Pr(>F) Pr(>F) given phy  
  # group      2 1.76960 0.88480   121.9 1.0963e-23           0.0351 *
  #   Residuals 71 0.51537 0.00726                                      
  # ---
  #   Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
  
  
  #phyl.aov.geiger<-aov.phylo(y.constrained~x.constrained.factor, prunetree.constrained, nsim=10000)
}


# 
# #try looping through randomized group assignments
# tmp<-as.data.frame(x)
# tmp$random<-tmp[,1]
# tmp$random<-sample(tmp$random)
# tmp$x <- NULL
# tmp<-tmp$random
# names(tmp)<-names(x.factor)
# phyl.aov.geiger<-aov.phylo(y~tmp, prunetree.shallow.time, nsim=1)
# #summary(phyl.aov.geiger)
# 
# #post hoc test
# require(multcomp)
# multCompTukey <- glht(phyl.aov.geiger, linfct = mcp(group = "Tukey"))
# summary(multCompTukey) #this outputs the p alues for each comparison
# #fix(aov.phylo)
# 


#running the generalized least squares version instead of anova (but it's the same thing)
#setting up datastets

{
  library(ape)
  library(nlme)
  library(geiger)
  
  pglsdata <-data.frame(species=y.constrained, r2t=y.constrained, habitat=(x.constrained), r2t.unconstrained=y.unconstrained, size=z.constrained, diaphragm = dia.constrained, palp = palp.constrained)#, spread=spread
  pglsdata$species<-rownames(pglsdata)
  pglsdata<-pglsdata[order(pglsdata$habitat),] #sort by habitat
  pglsdata$r2t.log<-log(pglsdata$r2t)
  pglsdata$size.log<-log(pglsdata$size)
  pglsdata$diaphragm.log<-log(pglsdata$diaphragm)
  pglsdata$palp.log <- log(pglsdata$palp)
  pglsdata$r2t.unconstrained.log<-log(pglsdata$r2t.unconstrained)
  #sort by tree labels
  pglsdata<-pglsdata[prunetree.constrained.time$tip.label,]
  
  #add colors
  pglsdata$cols <- pglsdata$habitat
  pglsdata$cols[pglsdata$cols=="Seep"]<-"#CAE3FC"#"#ece7f2"#"#1A48DC"#"#a1dab4"#"blue"
  pglsdata$cols[pglsdata$cols=="OF"]<-  "#007AC1"#"#2b8cbe"#1BD7DE"#"#225ea8"#"lightblue"
  pglsdata$cols[pglsdata$cols=="Vent"]<-"#00BCDF"#"#a6bddb"#42B649"#"#41b6c4"#grey"
  
  #add pch indicators for cases where size is missing
  pglsdata$pch <- ifelse(is.na(pglsdata$size),  23 , 21)
  
  #fill in missing sizes with BM
  size_recon<-pglsdata[,c("species", "size", "diaphragm", "palp")]
  size_recon$size<-log(size_recon$size)
  size_recon$diaphragm<-log(size_recon$diaphragm)
  size_recon$palp<-log(size_recon$palp)
  
  #univariate BM reconstruction
  size_recon1<-phylopars(trait_data=size_recon[,c(1:2)], tree=prunetree.constrained.time)$anc_recon
  size_recon1<-size_recon1[1:length(prunetree.constrained.time$tip.label)]
  size_recon1<-setNames(size_recon1, prunetree.constrained.time$tip.label)
  size_recon1<-exp(size_recon1)
  
  #attemtping multivariate BM reconstruction
  size_recon2.fit<-phylopars(trait_data=size_recon, tree=prunetree.constrained.time, model="BM")
  size_recon2.1.fit<-phylopars(trait_data=size_recon, tree=force.ultrametric(prunetree.constrained.time), model="mvOU", usezscores = FALSE)
  #size_recon2.2<-phylo.impute(tree=prunetree.constrained.time, X = as.matrix(size_recon[,-1]))$anc_recon
  AIC(size_recon2.fit) #293.3565
  AIC(size_recon2.1.fit) #297.2497
  
  require(mvMORPH)
  size_recon2.fit.mvbm<- mvMORPH::mvBM(tree=prunetree.constrained.time, data = as.matrix(size_recon[,-1]), model='BM1', param=list(root="stationary"))
  size_recon2.1.fit.mvou<-mvMORPH::mvOU(tree=prunetree.constrained.time, data = as.matrix(size_recon[,-1]), model='OU1', param=list(root="stationary"))
  size_recon2.fit.mvbm.2<- mvMORPH::mvBM(tree=prunetree.constrained.time, data = as.matrix(size_recon[,-1]), model='BM1')
  size_recon2.1.fit.mvou.2<-mvMORPH::mvOU(tree=prunetree.constrained.time, data = as.matrix(size_recon[,-1]), model='OU1')
  
  AIC(size_recon2.fit.mvbm) #308.0544
  AIC(size_recon2.1.fit.mvou) #277.3068
  AIC(size_recon2.fit.mvbm.2) #308.0544
  AIC(size_recon2.1.fit.mvou.2) #277.1937
  
  t1<-estim(tree=prunetree.constrained.time, data=as.matrix(size_recon[,-1]), object=size_recon2.fit.mvbm.2)
  t2<-estim(tree=prunetree.constrained.time, data=as.matrix(size_recon[,-1]), object=size_recon2.1.fit.mvou.2)
  
  plot(t1$estimates ~ t2$estimates)
  
  size_recon2<-as.data.frame(size_recon2.fit$anc_recon[1:length(prunetree.constrained.time$tip.label),])
  size_recon2<-setNames(size_recon2$size, prunetree.constrained.time$tip.label)
  size_recon2<-exp(size_recon2)
  
  size_recon2.1<-as.data.frame(size_recon2.1.fit$anc_recon[1:length(prunetree.constrained.time$tip.label),])
  size_recon2.1<-setNames(size_recon2.1$size, prunetree.constrained.time$tip.label)
  size_recon2.1<-exp(size_recon2.1)
  
  size_recon2.2<-exp(t1$estimates[,1])
  size_recon2.3<-exp(t2$estimates[,1])
  
  summary(lm(log(size_recon2) ~ log(size_recon1)))
  summary(lm(log(size_recon2.1) ~ log(size_recon1)))
  summary(lm(log(size_recon2.2) ~ log(size_recon1)))
  summary(lm(log(size_recon2.3) ~ log(size_recon1)))
  
  plot(log(size_recon2) ~ log(size_recon1))
  plot(log(size_recon2.1) ~ log(size_recon1))
  plot(log(size_recon2.2) ~ log(size_recon1))
  plot(log(size_recon2.3) ~ log(size_recon1))
  
  names(size_recon1) == prunetree.constrained.time$tip.label
  names(size_recon2) == prunetree.constrained.time$tip.label
  names(size_recon2.1) == prunetree.constrained.time$tip.label
  names(size_recon2.2) == prunetree.constrained.time$tip.label
  names(size_recon2.3) == prunetree.constrained.time$tip.label
  
  #size_recon <- size_recon1
  #size_recon <- size_recon2 #mvBM phylopars
  #size_recon <- size_recon2.1 #mvOU phylopars
  #size_recon <- size_recon2.2 #mvBM mvMORPH
  size_recon <- size_recon2.3 #mvOU mvMORPH
  
  #plot(log(size_recon2)~ log(size_recon2.1))
  
}

#generating supplementary data table output
recon_output<-data.frame(estimate = t2$estimates, variance = t2$var)
#write.table(recon_output, file='siboglinid_size.txt', quote = F, sep = '\t', row.names = T, col.names = T)

require(geiger)
anov.y<-log(setNames(pglsdata$r2t,rownames(pglsdata)))
anov.x<-as.factor(setNames(pglsdata$habitat,rownames(pglsdata)))
anov.z<-log(size_recon) #setNames(pglsdata$size,rownames(pglsdata))
nodes<-unlist(lapply(setNames(nodepath(congruify.constrained$phy), congruify.constrained$phy$tip.label), length))

phyl.aov.geiger.test<-aov.phylo(anov.y ~ anov.x, prunetree.constrained.time, nsim=1000)
attr(phyl.aov.geiger.test, "summary")
summary(phyl.aov.geiger.test)


#######
#stochastic character mapping
{
  habitat<-as.character(pglsdata$habitat)
  names(habitat)<-rownames(pglsdata)
  
  #setting up node labels
  cols<-setNames(palette()[1:length(unique(habitat))],sort(unique(habitat)))
  
  #modify colors to custom choices
  cols[2]<-"#CAE3FC"#"#ece7f2"#1A48DC"#"#a1dab4"#"blue"#<- "blue"
  cols[1]<-"#007AC1"#"#2b8cbe"#1BD7DE"#"#225ea8"#"lightblue"#<-"lightblue"
  cols[3]<-"#00BCDF"#"#a6bddb"#42B649"#"#41b6c4"#grey"#<-"grey"
        
  
  
  #plot(bind.tip(prunetree.shallow.time, tip.label="ROOT", edge.length=0.001, where=100, position=0))
  
  # #stochastic character1 map with custom model
  # TimeTree.simmap.habitat<-make.simmap(bind.tip(prunetree.shallow.time, tip.label="ROOT", edge.length=0.1, where=100, position=0), 
  #                                      c(habitat, ROOT="OF"), model='ARD', nsim=1000, message=T, type='discrete')#, pi=c(1,0,0))
  # 
  TimeTree.simmap.habitat<-make.simmap(prunetree.constrained.time, habitat, model='ARD', nsim=100, message=T, type='discrete',pi="fitzjohn")#, pi=c(1,0,0))
  
  
  #TimeTree.simmap.habitat.drop<-pbmcapply:::pbmclapply(TimeTree.simmap.habitat, drop.tip.simmap, tip="ROOT")
  #oldClass(TimeTree.simmap.habitat.drop)<-class(TimeTree.simmap.habitat)
  
  ARD.fit<-fitMk(prunetree.constrained.time, x=habitat, model='ARD')
  AIC(ARD.fit)
  
  dd<-density(TimeTree.simmap.habitat)
  plot(dd)
  par(new=T)
  par(mfrow=c(1,1),oma=c(9,9,9,9))
  plot(ARD.fit, show.zeros=T, cex.main=1, cex.traits=0.6, cex.rates=0.6)
  
  
  #posterior summary of stochastic mapping
  pd.ARD<- summary(TimeTree.simmap.habitat, plot=F)
  
  #move the node label
  tmp<-pd.ARD$ace
  
  #generate PDF plot
  #start plotting
  # 
  # h<-max(nodeHeights(TimeTree.simmap.habitat[[1]]))
  # #plotTree(TimeTree.simmap.habitat[[1]],plot=FALSE)
  # obj<-geo.legend(alpha=0.3,cex=1.2,plot=FALSE)
  # obj$leg<-h-obj$leg
  # plotSimmap(TimeTree.simmap.habitat[[1]], cols, fsize=0.35, ftype='i', lwd=2.5, direction='rightwards',ylim=c(-0.25*Ntip(TimeTree.simmap.habitat[[1]]),Ntip(TimeTree.simmap.habitat[[1]])))
  # #geo.legend(leg=obj$leg,colors=obj$colors,cex=1.2, alpha=0.2)
  # par(new=T)
}


#plot SIMMAP - constrained BLs
{
  pdf(file = "tubeworm.simmap-constrained_BL-nofren.pdf",
      width = 4,
      height = 8)
  
  #plotSimmap(TimeTree.simmap.habitat[[1]], cols, fsize=0.35, ftype='i', lwd=2.5, direction='rightwards')#,ylim=c(-0.25*Ntip(TimeTree.simmap.habitat[[1]]),Ntip(TimeTree.simmap.habitat[[1]])))
  #plotSimmap(TimeTree.simmap.habitat[[1]], cols, fsize=0.35, ftype='i', lwd=2.5, direction='rightwards')#,ylim=c(-0.25*Ntip(TimeTree.simmap.habitat[[1]]),Ntip(TimeTree.simmap.habitat[[1]])))
  
  #conmap<-contMap(tree = as.phylo(TimeTree.simmap.habitat.drop[[1]]), x = setNames((pglsdata$r2t.log), rownames(pglsdata)), plot=F)
  #get rate estimates
  #anc_rates<-getRates(trees=c(TimeTree.simmap.habitat.drop[[1]], prunetree.shallow))
  anc_sub_site <-
    get_matched_edgelengths(trees = c(as.phylo(TimeTree.simmap.habitat[[1]]), prunetree.constrained))[, 2]
  
  tiprates<-get_matched_edgelengths(trees = c(as.phylo(TimeTree.simmap.habitat[[1]]), prunetree.constrained))
  tiprates<-tiprates[,2]/tiprates[,1]
  #tiprates<-tiprates#[1:74]
  #tiprates<-log(tiprates)
  #tiprates <- tiprates[tiprates > quantile(tiprates, c(0.05, 0.95)) [1] & tiprates < quantile(tiprates, c(0.05, 0.95)) [2]]
  #max(exp(tiprates[1:74]))/min(exp(tiprates[1:74]))
  
  #tiprates2<-terminalLengths(prunetree.shallow)[TimeTree.simmap.habitat.drop[[1]]$tip.label]/terminalLengths(TimeTree.simmap.habitat.drop[[1]])
  #tiprates seem correct
  
  #trying to get the internal branch lengths from the phylogram to plot inside the contmap
  #not working currently
  #  {#plot(prunetree.shallow, cex=0.5)
  # # nodelabels(node =
  # #              node_indices_edge(tree = prunetree.shallow, edges = c(1:length(prunetree.shallow$edge.length)))[!
  # #              node_indices_edge(tree = prunetree.shallow, edges = c(1:length(prunetree.shallow$edge.length))) %in% 1:99]
  # #          )
  # 
  #  a<-setNames(
  #    prunetree.shallow$edge.length,
  #    node_indices_edge(tree = prunetree.shallow, edges = c(
  #      1:length(prunetree.shallow$edge.length)
  #    )))
  #  a<-a[!as.numeric(names(a)) %in% 1:99]
  #  a<-setNames(a, match_nodes(t1=prunetree.shallow, t2=TimeTree.simmap.habitat.drop[[1]])[,2][-1])
  #  
  # conmap <-
  #     contMap(
  #       tree = as.phylo(TimeTree.simmap.habitat.drop[[1]]),
  #       x =
  #         terminalLengths(prunetree.shallow),
  #         plot = F, method='user',
  #       anc.states = a)
  # }
  
  conmap <-
    contMap(
      tree = as.phylo(TimeTree.simmap.habitat[[1]]),
      x = (anc_sub_site[1:74]),
      plot = F, method='user',
      anc.states = (anc_sub_site[-c(1:74)]))
  
  #conmap<-setMap(conmap,c("white","#FFFFB2","#FECC5C","#FD8D3C","#E31A1C"))
  conmap <-
    setMap(
      conmap,
      #rev(c(viridis::viridis(8)))[-1],
      c(
        "#f0f0f0",
                 "#d9d9d9",
                 "#bdbdbd",
                 "#969696",
                 "#737373",
                 "#525252",
                 "#252525",
                 "#000000"
      ),
      bias = 1.35,
      interpolate = c("spline")
    )
  
  tree<-as.phylo(TimeTree.simmap.habitat[[1]])
  # dotTree(tree,habitat,colors=cols,lwd=3.5,fsize=0.35, ftype='i', sig=2)
  # par(fg="transparent")
  # 
  # plot(conmap$tree,add=TRUE,lwd=3.5,colors=conmap$cols,
  #      ylim=c(-1/25*Ntip(tree),Ntip(tree)),offset=1.7)
  # 
  # par(fg="black")
  # 
  # add.color.bar(0.3*max(nodeHeights(tree)),conmap$cols,title="rate",
  #               lims=conmap$lims,digits=3,prompt=FALSE,x=0.3*max(nodeHeights(conmap)),
  #               y=0.4*(1+par()$usr[3]),lwd=4,fsize=1,subtitle="")
  
  
  #layout(matrix(c(1,2),1,2))
  #plot(1, type="n", xlab="", ylab="", xlim=c(0, 10), ylim=c(0, 10), bty='n')
  par(oma=c(0,0,0,0))
  
  plot(
    conmap,
    outline = F,
    fsize = 0.35,
    ftype = 'i',
    lwd = 3.0,
    sig = 3,
    offset = 0.65,
    leg.txt="scaled rate"
  )
  
  par(xpd=T)
  axisPhylo(pos=-7, cex.axis=0.5)
  
  # points(
  #   x = rep(max(nodeHeights(tree)) + 2, Ntip(tree)),
  #   y = seq(1, Ntip(tree), length.out = Ntip(tree)),
  #   pch = 21,
  #   bg = pglsdata[conmap$tree$tip.label, ]$cols,
  #   lwd = 0.5
  # )
  #-1/(Ntip(tree))*Ntip(tree)
  #add node labels
  par(lwd = 0.1)
  
  #relabel to get correct node assignments
  
  # rownames(tmp) <-
  #   c(matchNodes(tr1 = prunetree.shallow.time, tr2 = conmap$tree)[, 2],
  #     rownames(tmp)[c(Ntip(tree):length(rownames(tmp)))])
  # tmp$names<-rownames(tmp)
  #
  
  #dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1
  #tmp[c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1),]
  
  
  #plot(tree,cex=0.1)
  
  
  # nodelabels(node = as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))]),
  #   pie = tmp[as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))])-99],
  #   piecol = scales::alpha(cols, 1.0),
  #   prompt = FALSE,
  #   cex = 0.6
  # )
  
  nodelabels(#node = as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))]),
    pie = tmp,#[as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))])-99],
    piecol = scales::alpha(cols, 1.0),
    prompt = FALSE,
    cex = 0.6
  )
  
  tiplabels(
    pie = tmp[conmap$tree$tip.label,],
    piecol = scales::alpha(cols, 0.85),
    prompt = FALSE,
    cex = 0.3,
    offset=7.5
  )
  
  
  # barplot(rev(r2t.prunetree.constrained.deep[conmap$tree$tip.label, ]), horiz =
  #           T, ylim=c(11, 99))
  
  
  # plotTree.barplot(conmap$tree,log(setNames(r2t.prunetree.constrained.deep[conmap$tree$tip.label,], conmap$tree$tip.label)),
  #                  args.plotTree=list(plot=FALSE),
  #                  args.barplot=list(xlab = expression(paste("tip s/s/Ma"))),
  #                  add=TRUE,ylim=c(3, 161),cex.lab=0.9,cex.axis=0.9)
  
  # plotTree.barplot(conmap$tree,tiprates[conmap$tree$tip.label],
  #                  args.plotTree=list(plot=FALSE),
  #                  args.barplot=list(xlab = expression(paste("tip s/s/Ma"))),
  #                  add=TRUE,ylim=c(3, 161),cex.lab=0.9,cex.axis=0.9)
  
  # #extract tip states
  # tipcols<-pglsdata[prunetree.shallow.time$tip.label,]$cols
  # 
  # #add tip labels # can only be executed after pars recon for this sectino
  # tiplabels(pch=20, col=tipcols, cex=0.8)
  
  #add legend
  #add.simmap.legend(colors=cols,fsize=0.75, x=15, y=80, prompt=F, shape="circle", cex=2)
  
  legend(x=12, y=50, legend = names(cols)[c(1,3,2)], pt.bg=cols[c(1,3,2)], pch=21, cex=1, pt.cex=2, bty='n')
  
  dev.off()
}


#plot SIMMAP - unconstrained BLs #DOES NOT WORK WITH EDGE MATCHING CURRENTLY
{
  pdf(file = "tubeworm.simmap-unconstrained_BL-nofren.pdf",
      width = 4,
      height = 8)
  
  #plotSimmap(TimeTree.simmap.habitat[[1]], cols, fsize=0.35, ftype='i', lwd=2.5, direction='rightwards')#,ylim=c(-0.25*Ntip(TimeTree.simmap.habitat[[1]]),Ntip(TimeTree.simmap.habitat[[1]])))
  #plotSimmap(TimeTree.simmap.habitat[[1]], cols, fsize=0.35, ftype='i', lwd=2.5, direction='rightwards')#,ylim=c(-0.25*Ntip(TimeTree.simmap.habitat[[1]]),Ntip(TimeTree.simmap.habitat[[1]])))
  
  #conmap<-contMap(tree = as.phylo(TimeTree.simmap.habitat.drop[[1]]), x = setNames((pglsdata$r2t.log), rownames(pglsdata)), plot=F)
  #get rate estimates
  #anc_rates<-getRates(trees=c(TimeTree.simmap.habitat.drop[[1]], prunetree.shallow))
  anc_sub_site <-
    get_matched_edgelengths(trees = c(as.phylo(TimeTree.simmap.habitat[[1]]), prunetree.unconstrained))[, 2]
  
  tiprates<-get_matched_edgelengths(trees = c(TimeTree.simmap.habitat[[1]], prunetree.unconstrained))
  tiprates<-tiprates[,2]/tiprates[,1]
  
  #tiprates2<-terminalLengths(prunetree.shallow)[TimeTree.simmap.habitat.drop[[1]]$tip.label]/terminalLengths(TimeTree.simmap.habitat.drop[[1]])
  #tiprates seem correct
  
  #trying to get the internal branch lengths from the phylogram to plot inside the contmap
  #not working currently
  #  {#plot(prunetree.shallow, cex=0.5)
  # # nodelabels(node =
  # #              node_indices_edge(tree = prunetree.shallow, edges = c(1:length(prunetree.shallow$edge.length)))[!
  # #              node_indices_edge(tree = prunetree.shallow, edges = c(1:length(prunetree.shallow$edge.length))) %in% 1:99]
  # #          )
  # 
  #  a<-setNames(
  #    prunetree.shallow$edge.length,
  #    node_indices_edge(tree = prunetree.shallow, edges = c(
  #      1:length(prunetree.shallow$edge.length)
  #    )))
  #  a<-a[!as.numeric(names(a)) %in% 1:99]
  #  a<-setNames(a, match_nodes(t1=prunetree.shallow, t2=TimeTree.simmap.habitat.drop[[1]])[,2][-1])
  #  
  # conmap <-
  #     contMap(
  #       tree = as.phylo(TimeTree.simmap.habitat.drop[[1]]),
  #       x =
  #         terminalLengths(prunetree.shallow),
  #         plot = F, method='user',
  #       anc.states = a)
  # }
  
  conmap <-
    contMap(
      tree = as.phylo(TimeTree.simmap.habitat[[1]]),
      x = (anc_sub_site[1:74]),
      plot = F, method='user',
      anc.states = (anc_sub_site[-c(1:74)]))
  
  #conmap<-setMap(conmap,c("white","#FFFFB2","#FECC5C","#FD8D3C","#E31A1C"))
  conmap <-
    setMap(
      conmap,
      #rev(c(viridis::viridis(8)))[-1],
      c(
        "#f0f0f0",
                 "#d9d9d9",
                 "#bdbdbd",
                 "#969696",
                 "#737373",
                 "#525252",
                 "#252525",
                 "#000000"
      ),
      bias = 1.35,
      interpolate = c("spline")
    )
  
  tree<-as.phylo(TimeTree.simmap.habitat[[1]])
  # dotTree(tree,habitat,colors=cols,lwd=3.5,fsize=0.35, ftype='i', sig=2)
  # par(fg="transparent")
  # 
  # plot(conmap$tree,add=TRUE,lwd=3.5,colors=conmap$cols,
  #      ylim=c(-1/25*Ntip(tree),Ntip(tree)),offset=1.7)
  # 
  # par(fg="black")
  # 
  # add.color.bar(0.3*max(nodeHeights(tree)),conmap$cols,title="rate",
  #               lims=conmap$lims,digits=3,prompt=FALSE,x=0.3*max(nodeHeights(conmap)),
  #               y=0.4*(1+par()$usr[3]),lwd=4,fsize=1,subtitle="")
  
  
  #layout(matrix(c(1,2),1,2))
  #plot(1, type="n", xlab="", ylab="", xlim=c(0, 10), ylim=c(0, 10), bty='n')
  par(oma=c(0,0,0,0))
  
  plot(
    conmap,
    outline = F,
    fsize = 0.35,
    ftype = 'i',
    lwd = 3.0,
    sig = 3,
    offset = 0.65,
    leg.txt="scaled rate"
  )
  
  par(xpd=T)
  axisPhylo(pos=-7, cex.axis=0.5)
  
  # points(
  #   x = rep(max(nodeHeights(tree)) + 2, Ntip(tree)),
  #   y = seq(1, Ntip(tree), length.out = Ntip(tree)),
  #   pch = 21,
  #   bg = pglsdata[conmap$tree$tip.label, ]$cols,
  #   lwd = 0.5
  # )
  #-1/(Ntip(tree))*Ntip(tree)
  #add node labels
  par(lwd = 0.1)
  
  #relabel to get correct node assignments
  
  # rownames(tmp) <-
  #   c(matchNodes(tr1 = prunetree.shallow.time, tr2 = conmap$tree)[, 2],
  #     rownames(tmp)[c(Ntip(tree):length(rownames(tmp)))])
  # tmp$names<-rownames(tmp)
  #
  
  #dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1
  #tmp[c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1),]
  
  
  #plot(tree,cex=0.1)
  
  
  # nodelabels(node = as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))]),
  #   pie = tmp[as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))])-99],
  #   piecol = scales::alpha(cols, 1.0),
  #   prompt = FALSE,
  #   cex = 0.6
  # )
  
  nodelabels(#node = as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))]),
    pie = tmp,#[as.numeric(rownames(tmp)[which(c(dispRity::tree.age(as.phylo(conmap$tree))[-c(1:99),][,1] > 1))])-99],
    piecol = scales::alpha(cols, 1.0),
    prompt = FALSE,
    cex = 0.6
  )
  
  tiplabels(
    pie = tmp[conmap$tree$tip.label,],
    piecol = scales::alpha(cols, 0.85),
    prompt = FALSE,
    cex = 0.3,
    offset=7.5
  )
  
  
  # barplot(rev(r2t.prunetree.constrained.deep[conmap$tree$tip.label, ]), horiz =
  #           T, ylim=c(11, 99))
  
  
  # plotTree.barplot(conmap$tree,log(setNames(r2t.prunetree.constrained.deep[conmap$tree$tip.label,], conmap$tree$tip.label)),
  #                  args.plotTree=list(plot=FALSE),
  #                  args.barplot=list(xlab = expression(paste("tip s/s/Ma"))),
  #                  add=TRUE,ylim=c(3, 161),cex.lab=0.9,cex.axis=0.9)
  
  # plotTree.barplot(conmap$tree,tiprates[conmap$tree$tip.label],
  #                  args.plotTree=list(plot=FALSE),
  #                  args.barplot=list(xlab = expression(paste("tip s/s/Ma"))),
  #                  add=TRUE,ylim=c(3, 161),cex.lab=0.9,cex.axis=0.9)
  
  # #extract tip states
  # tipcols<-pglsdata[prunetree.shallow.time$tip.label,]$cols
  # 
  # #add tip labels # can only be executed after pars recon for this sectino
  # tiplabels(pch=20, col=tipcols, cex=0.8)
  
  #add legend
  #add.simmap.legend(colors=cols,fsize=0.75, x=15, y=80, prompt=F, shape="circle", cex=2)
  
  legend(x=12, y=50, legend = names(cols)[c(1,3,2)], pt.bg=cols[c(1,3,2)], pch=21, cex=1, pt.cex=2, bty='n')
  
  dev.off()
}


#set up phylomorphospace

#some plotting
o<-ordered(pglsdata$habitat, levels=c('OF', 'Vent', 'Seep'))
boxplot(r2t.log~o,data =pglsdata, notch=F)
#boxplot(x=pglsdata[pglsdata$habitat=='OF',]$r2t, y=pglsdata[pglsdata$habitat=='Vent',]$r2t)


# #plotting phylomorphospace - constrained
# {
#   Z<-as.data.frame(as.matrix(phylopars(cbind(pglsdata[1],log(pglsdata[5])),prunetree.constrained.time)$anc_recon[1:74,]))
#   #Z$r2t<-pglsdata[2]
#   Z<-merge(Z, as.data.frame(pglsdata[6]), by='row.names', all=TRUE)
#   colnames(Z)<- c('species', 'log(size)', 'r2t')
#   rownames(Z)<- Z$species
#   Z$species<-NULL
#   
#   pdf(file="tubeworm.phylomorphospace_constrained_BL.pdf", width=6, height=7)
#   phylomorphospace(prunetree.constrained.time, X=Z, label='off', ylim=c(-0.65,0.2), xlim=c(1,8),node.size=c(0.0001,1), lwd=0.15, bty='n', ylab="log(r2t)")#, control=c(col.edge='white'))
#   
#   title(main='tubeworm phylomorphospace')
#   #par(new=T)
#   #plot(r2t.log~log(size), col = pglsdata$cols, data = pglsdata, pch=16, ylim=c(-0.55,0.15), xlim=c(1.5,6), bty='n', xaxt='n', yaxt='n', ylab="", xlab="")
#   points(r2t~`log(size)`, bg = pglsdata$cols, data = Z, pch=21, lwd=0.5)
#   
#   #points(r2t~`log(size)`, bg = make.transparent("white",0.5), data = Z[which(is.na(pglsdata[5]$size)),], pch=21, lwd=0.5, cex=1)
#   points(r2t~`log(size)`, col = "white", data = Z[which(is.na(pglsdata[5]$size)),], pch=19, lwd=0.5, cex=0.5)
#   
#   
#   abline(lm(r2t.log~log(size), data =pglsdata), lwd=2, lty=1, col=make.transparent("red", 0.5))
#   abline(phylopars.lm(r2t.log ~ log(size), tree=prunetree.constrained.time, trait_data=pglsdata, model='BM'), col=make.transparent('red',0.5), lwd=4, lty=1)
#   
#   #summary(phylopars.lm(r2t.log ~ log(size), tree=prunetree.shallow.time, trait_data=pglsdata, model='BM'))
#   
#   #summary(lm(log(r2t)~log(size), data =pglsdata))
#   legend('topleft', title='habitat type', c('Organic fall','Seep', 'Vent'), col=c('blue', 'lightblue', 'grey'), pch=16, bty='n')
#   
#   #legend('top', text.col='red', legend=c('PGLS p~0.97'), cex=0.85, box.lwd=0.5) 
#   #legend('topleft', text.col='black', legend=c('LM R^2=0.328, p<<0.05'), cex=0.85, box.lwd=0.5)
#   dev.off()
# }


#plotting phylomorphospace - constrained
{
  tmp_nodecounts<-unlist(lapply(setNames(nodepath(prunetree.constrained.time), prunetree.constrained.time$tip.label), length))[prunetree.constrained.time$tip.label]
  Z<-cbind(as.matrix(log(size_recon)), tmp_nodecounts[prunetree.constrained.time$tip.label]) #,tiprates[prunetree.shallow.time$tip.label]
  #Z$r2t<-pglsdata[2]
  Z<-merge(Z, as.data.frame(log(pglsdata[2])), by='row.names', all=TRUE)
  colnames(Z)<- c('species', 'log.size','nodes','log.r2t')
  rownames(Z)<- Z$species
  Z$species<-NULL
  Z<- Z[rownames(pglsdata),]
  Z.names<-Z
  Z.names$species <- rownames(Z.names)
  
  pdf(file="tubeworm.phylomorphospace_constrained_bls.pdf", width=6, height=7)
  phylomorphospace(prunetree.constrained.time, X=Z[,c(-2)], label='off', ylim=c(-0.6,0.05), xlim=c(1,8),node.size=c(0.0000,0), lwd=0.35, bty='n', ylab="log(r2t)",xlab="log(size)", fsize=0.5)#, control=c(col.edge='white'))
  
  title(main='tubeworm phylomorphospace')
  #par(new=T)
  #plot(r2t.log~log(size), col = pglsdata$cols, data = pglsdata, pch=16, ylim=c(-0.55,0.15), xlim=c(1.5,6), bty='n', xaxt='n', yaxt='n', ylab="", xlab="")
  points(log.r2t~log.size, bg = scales::alpha(pglsdata$cols, 0.95), data = Z[rownames(pglsdata),], pch=pglsdata$pch, lwd=1, cex=1.25)  ## fix here
  
  #points(r2t~`log(size)`, col = "black", data = Z[which(is.na(pglsdata[4]$size)),], pch=19, lwd=0.5, cex=0.5)
  #points(r2t~`log(size)`, col = "black", data = Z, pch=19, lwd=0.5, cex=0.15)
  
  
  #add the bootstrap lines
  # for(i in c(1:2)) {
  #   cc <- (model$bootconfint95[i,])
  #   abline(a=cc[1], b=cc[2], lwd=1, col="red")
  #   #curve(plogis(cc[1]+cc[2]*x),col=make.transparent(palette.colors(palette = "Okabe-Ito")[2], 0.25),add=TRUE, lwd=0.25, col='red')
  # }
  
  model<-gls(log.r2t ~ log.size, correlation = corBrownian(phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  model.int<-gls(log.r2t ~ 1, correlation = corBrownian(phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  
  #res<-caper::residuals.pgls(model, phylo = TRUE)
  #res<- res/sqrt(var(res))[1] #standardises residuals by sqrt of their variance
  #rownames(res)[(abs(res)>3)]#gives the names of the outliers
  
  AIC(model)
  AIC(model.int)
  
  abline(model, col=make.transparent('black',1), lwd=4, lty=2)
  pGLS_ci<-gls.ci(Z$log.r2t,Z$log.size,vcv(prunetree.constrained.time))
  lines(pGLS_ci$CI.plot$X,pGLS_ci$CI.plot$Lower2.5,lty=2)
  lines(pGLS_ci$CI.plot$X,pGLS_ci$CI.plot$Upper2.5,lty=2)
  
  
  ### this add the labels to the points using their rownames
  ### font = 2 is bold
  
  #text(log.r2t ~ log.size, labels=rownames(Z[rownames(pglsdata),]),data=Z.names, cex=0.4, font=2)
  
  
  #model.OU <- gls(log.r2t ~ log.size, correlation = corMartins(1, phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  #AIC(model.OU)
  #summary(model.OU)
  #abline(model.OU, col=make.transparent('red',1), lwd=4, lty=2)

  #model.pagel <- gls(log.r2t ~ log.size, correlation = corPagel(0.5, phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  #IC(model.pagel)
  #summary(model.OU)
  #abline(model.pagel, col=make.transparent('red',1), lwd=4, lty=2)

  #tmp<-corMartins(0.00122085, phy = prunetree.constrained.time, form = ~species)
  #str(tmp)
  
  
  #pGLS_pi<-gls.pi(Z$log.r2t,Z$log.size,vcv(prunetree.shallow.time),1000)
  #lines(pGLS_pi$PI.plot$X,pGLS_pi$PI.plot$Lower2.5,lty=3)
  #lines(pGLS_pi$PI.plot$X,pGLS_pi$PI.plot$Upper2.5,lty=3)
  
  #abline(lm(r2t.log~log(size_recon), data = pglsdata), lwd=2, lty=1, col=make.transparent("red", 0.5))
  #model<-phylopars.lm(r2t.log ~ log(size), tree=prunetree.shallow.time, trait_data=pglsdata, model='BM')
  #abline(model, col=make.transparent('black',1), lwd=4, lty=2)
  
  #summary(phylopars.lm(r2t.log ~ log(size), tree=prunetree.shallow.time, trait_data=pglsdata, model='BM'))
  
  #summary(lm(log(r2t)~log(size), data =pglsdata))
  legend('topright', title='habitat type', c('Organic fall','Seep', 'Vent')[c(1,3,2)], col=cols[c(1,3,2)], pch=16, bty='n', cex=1)
  
  #legend('top', text.col='red', legend=c('PGLS p~0.97'), cex=0.85, box.lwd=0.5) 
  #legend('topleft', text.col='black', legend=c('LM R^2=0.328, p<<0.05'), cex=0.85, box.lwd=0.5)
  dev.off()
  
}


tmp<-phylolm::phylolm(log.r2t ~ log.size, data = Z.names[,c(-4)],  phy= prunetree.constrained.time, boot=1000)
plot(tmp$residuals ~ log.size, data = Z.names[,c(-4)])
plot(tmp)
qqnorm(model,abline = c(0,1))
predictmeans::residplot(model)
tmp$bootconfint95


#trying with only taxa with non-missing data
tmp<-phylolm::phylolm(r2t.log ~ size.log, data = pglsdata,  phy= prunetree.constrained.time, boot=1000)
#plot(tmp$residuals ~ size.log, data = pglsdata)
plot(tmp)
qqnorm(model,abline = c(0,1))
predictmeans::residplot(model)
tmp$bootconfint95


#trying independent sister pairs
require(diverge)
require(boot)
sisters<-analyze_sister_pairs(tree = prunetree.constrained.time, data = Z.names, trait1 = 'log.r2t', trait2='log.size', useAbsolute = F)
#run the sister pair analysis on constrained and unconstrained datasets

#plotting phylomorphospace - unconstrained
{
  tmp_nodecounts<-unlist(lapply(setNames(nodepath(prunetree.constrained.time), prunetree.constrained.time$tip.label), length))[prunetree.constrained.time$tip.label]
  Z<-cbind(as.matrix(log(size_recon)), tmp_nodecounts[prunetree.constrained.time$tip.label]) #,tiprates[prunetree.shallow.time$tip.label]
  #Z$r2t<-pglsdata[2]
  Z<-merge(Z, as.data.frame(log(pglsdata[4])), by='row.names', all=TRUE)
  colnames(Z)<- c('species', 'log.size','nodes','log.r2t')
  rownames(Z)<- Z$species
  Z$species<-NULL
  Z<- Z[rownames(pglsdata),]
  Z.names<-Z
  Z.names$species <- rownames(Z.names)
  
  pdf(file="tubeworm.phylomorphospace_unconstrained_bls.pdf", width=6, height=7)
  phylomorphospace(prunetree.constrained.time, X=Z[,c(-2)], label='off', ylim=c(-0.7,0.05), xlim=c(1,8),node.size=c(0.0000,0), lwd=0.35, bty='n', ylab="log(r2t)",xlab="log(size)", fsize=0.5)#, control=c(col.edge='white'))
  
  title(main='tubeworm phylomorphospace')
  #par(new=T)
  #plot(r2t.log~log(size), col = pglsdata$cols, data = pglsdata, pch=16, ylim=c(-0.55,0.15), xlim=c(1.5,6), bty='n', xaxt='n', yaxt='n', ylab="", xlab="")
  points(log.r2t~log.size, bg = scales::alpha(pglsdata$cols, 0.95), data = Z[rownames(pglsdata),], pch=pglsdata$pch, lwd=1, cex=1.25)  ## fix here
  
  #points(r2t~`log(size)`, col = "black", data = Z[which(is.na(pglsdata[4]$size)),], pch=19, lwd=0.5, cex=0.5)
  #points(r2t~`log(size)`, col = "black", data = Z, pch=19, lwd=0.5, cex=0.15)
  
  
  #add the bootstrap lines
  # for(i in c(1:2)) {
  #   cc <- (model$bootconfint95[i,])
  #   abline(a=cc[1], b=cc[2], lwd=1, col="red")
  #   #curve(plogis(cc[1]+cc[2]*x),col=make.transparent(palette.colors(palette = "Okabe-Ito")[2], 0.25),add=TRUE, lwd=0.25, col='red')
  # }
  
  model<-gls(log.r2t ~ log.size, correlation = corBrownian(phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  model.int<-gls(log.r2t ~ 1, correlation = corBrownian(phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  
  #res<-caper::residuals.pgls(model, phylo = TRUE)
  #res<- res/sqrt(var(res))[1] #standardises residuals by sqrt of their variance
  #rownames(res)[(abs(res)>3)]#gives the names of the outliers
  
  AIC(model)
  AIC(model.int)
  
  abline(model, col=make.transparent('black',1), lwd=4, lty=2)
  pGLS_ci<-gls.ci(Z$log.r2t,Z$log.size,vcv(prunetree.constrained.time))
  lines(pGLS_ci$CI.plot$X,pGLS_ci$CI.plot$Lower2.5,lty=2)
  lines(pGLS_ci$CI.plot$X,pGLS_ci$CI.plot$Upper2.5,lty=2)
  
  
  ### this add the labels to the points using their rownames
  ### font = 2 is bold
  
  #text(log.r2t ~ log.size, labels=rownames(Z[rownames(pglsdata),]),data=Z.names, cex=0.4, font=2)
  
  
  #model.OU <- gls(log.r2t ~ log.size, correlation = corMartins(1, phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  #AIC(model.OU)
  #summary(model.OU)
  #abline(model.OU, col=make.transparent('red',1), lwd=4, lty=2)
  
  #model.pagel <- gls(log.r2t ~ log.size, correlation = corPagel(0.5, phy = prunetree.constrained.time, form = ~species), data = Z.names, method = "ML")
  #IC(model.pagel)
  #summary(model.OU)
  #abline(model.pagel, col=make.transparent('red',1), lwd=4, lty=2)
  
  #tmp<-corMartins(0.00122085, phy = prunetree.constrained.time, form = ~species)
  #str(tmp)
  
  
  #pGLS_pi<-gls.pi(Z$log.r2t,Z$log.size,vcv(prunetree.shallow.time),1000)
  #lines(pGLS_pi$PI.plot$X,pGLS_pi$PI.plot$Lower2.5,lty=3)
  #lines(pGLS_pi$PI.plot$X,pGLS_pi$PI.plot$Upper2.5,lty=3)
  
  #abline(lm(r2t.log~log(size_recon), data = pglsdata), lwd=2, lty=1, col=make.transparent("red", 0.5))
  #model<-phylopars.lm(r2t.log ~ log(size), tree=prunetree.shallow.time, trait_data=pglsdata, model='BM')
  #abline(model, col=make.transparent('black',1), lwd=4, lty=2)
  
  #summary(phylopars.lm(r2t.log ~ log(size), tree=prunetree.shallow.time, trait_data=pglsdata, model='BM'))
  
  #summary(lm(log(r2t)~log(size), data =pglsdata))
  legend('topright', title='habitat type', c('Organic fall','Seep', 'Vent')[c(1,3,2)], col=cols[c(1,3,2)], pch=16, bty='n', cex=1)
  
  #legend('top', text.col='red', legend=c('PGLS p~0.97'), cex=0.85, box.lwd=0.5) 
  #legend('topleft', text.col='black', legend=c('LM R^2=0.328, p<<0.05'), cex=0.85, box.lwd=0.5)
  dev.off()
  
}


#plotting phylomorphospace - unconstrained
# {
#   Z<-as.data.frame(as.matrix(phylopars(cbind(pglsdata[1],log(pglsdata[5])),prunetree.constrained.time)$anc_recon[1:74,]))
#   #Z$r2t<-pglsdata[2]
#   Z<-merge(Z, as.data.frame(log(pglsdata[4])), by='row.names', all=TRUE)
#   colnames(Z)<- c('species', 'log(size)', 'r2t')
#   rownames(Z)<- Z$species
#   Z$species<-NULL
#   
#   pdf(file="tubeworm.phylomorphospace_unconstrained_BL.pdf", width=6, height=7)
#   phylomorphospace(prunetree.constrained.time, X=Z, label='off', ylim=c(-0.7,0.1), xlim=c(1,8),node.size=c(0.0001,1), lwd=0.15, bty='n', ylab="log(r2t)")#, control=c(col.edge='white'))
#   
#   title(main='tubeworm phylomorphospace')
#   #par(new=T)
#   #plot(r2t.log~log(size), col = pglsdata$cols, data = pglsdata, pch=16, ylim=c(-0.55,0.15), xlim=c(1.5,6), bty='n', xaxt='n', yaxt='n', ylab="", xlab="")
#   points(r2t~`log(size)`, bg = pglsdata$cols, data = Z, pch=21, lwd=0.5)
#   
#   #points(r2t~`log(size)`, bg = make.transparent("white",0.5), data = Z[which(is.na(pglsdata[5]$size)),], pch=21, lwd=0.5, cex=1)
#   points(r2t~`log(size)`, col = "white", data = Z[which(is.na(pglsdata[5]$size)),], pch=19, lwd=0.5, cex=0.5)
#   
#   
#   abline(lm(log(r2t.unconstrained)~log(size), data =pglsdata), lwd=2, lty=1, col=make.transparent("red", 0.5))
#   abline(phylopars.lm(log(r2t.unconstrained) ~ log(size), tree=prunetree.constrained.time, trait_data=pglsdata, model='BM'), col=make.transparent('red',0.5), lwd=4, lty=1)
#   
#   #summary(phylopars.lm(r2t.log ~ log(size), tree=prunetree.shallow.time, trait_data=pglsdata, model='BM'))
#   
#   #summary(lm(log(r2t)~log(size), data =pglsdata))
#   legend('topleft', title='habitat type', c('Organic fall','Seep', 'Vent'), col=c('blue', 'lightblue', 'grey'), pch=16, bty='n')
#   
#   #legend('top', text.col='red', legend=c('PGLS p~0.97'), cex=0.85, box.lwd=0.5) 
#   #legend('topleft', text.col='black', legend=c('LM R^2=0.328, p<<0.05'), cex=0.85, box.lwd=0.5)
#   dev.off()
# }



#
#retrying phylogenetic anova in a different way

require(geiger)
anov.y<-(setNames((log(pglsdata$r2t)),rownames(pglsdata)))
anov.x<-factor(setNames(pglsdata$habitat,rownames(pglsdata)), ordered=F)
# {
# anov.x<-as.character(pglsdata$habitat)
# anov.x[anov.x=="OF"]<-"bOF"
# anov.x[anov.x=="Vent"]<-"cVent"
# anov.x[anov.x=="Seep"]<-"aSeep"
# anov.x<-setNames(as.factor(anov.x), rownames(pglsdata))
# }
anov.z<-log(size_recon) #setNames(pglsdata$size,rownames(pglsdata))
nodes<-unlist(lapply(setNames(nodepath(prunetree.constrained.time), prunetree.constrained.time$tip.label), length))

phyl.aov.geiger.test<-aov.phylo(anov.y ~ anov.x, prunetree.constrained.time, nsim=1000)
attr(phyl.aov.geiger.test, "summary")


brmsdata<-data.frame(rate=anov.y, habitat=anov.x, size=anov.z, species=names(anov.y), nodecounts=log(nodes))

plot(brmsdata$rate~ brmsdata$nodecounts)


require(phylolm)
#set up for rr2 and phylolm -- only for constrained
{
  require(rr2)
  #trying with phylolm
  forms<-c("habitat",
           "habitat + nodecounts",
           "habitat + size + nodecounts",
           "habitat + size + nodecounts + habitat:size")
  
  starphylo <- ape::compute.brlen(prunetree.constrained.time, method = "Grafen", power = .0001)
  #ancova.bm1<-phylolm::phylolm(rate~ 0, data=brmsdata, phy=prunetree.shallow.time, model="BM", boot=100)
  #fitContinuous(phy=prunetree.shallow.time, dat=setNames(brmsdata$rate, rownames(brmsdata)), model="BM")
  #phylosig(prunetree.shallow.time, x=setNames(brmsdata$rate, rownames(brmsdata)), test=T, method="lambda")
  #summary(lm(brmsdata$rate ~ brmsdata$nodecounts))
  
  
  #test1<-untangle(rotate(prunetree.shallow.time, node=50), "read.tree")
  # ancova.bm1<-phylolm::phylolm(size~ habitat, data=brmsdata, phy=test, model="BM", boot=100)
  # summary(ancova.bm1)
  # ancova.bm1<-phylolm::phylolm(rateDR ~ habitat+ size + nodecounts + habitat:size, data=brmsdata, phy=prunetree.shallow.time, model="BM", boot=100)
  # summary(ancova.bm1)
  # plot(rateDR ~ habitat, data=brmsdata)
  
  ancova.bm1<-phylolm::phylolm(rate ~ habitat, data=brmsdata, phy=prunetree.constrained.time, model="BM", boot=100)
  ancova.bm1.r<-lm(rate~ habitat, data=brmsdata)
  ancova.bm1.star<-phylolm::phylolm(rate~ habitat, data=brmsdata, phy=starphylo, model="BM", boot=100)
  
  R2_lik(ancova.bm1, ancova.bm1.star)
  
  ancova.bm2<-phylolm::phylolm(rate ~ habitat + nodecounts, data=brmsdata, phy=prunetree.constrained.time, model="BM", boot=100)
  ancova.bm2.r<-lm(rate ~ habitat + nodecounts, data=brmsdata)
  ancova.bm2.star<-phylolm::phylolm(rate ~ habitat + nodecounts, data=brmsdata, phy=starphylo, model="BM", boot=100)
  
  R2_lik(ancova.bm2, ancova.bm1) #partial R2 for nodecounts
  
  ancova.bm3<-phylolm::phylolm(rate ~ habitat + size + nodecounts, data=brmsdata, phy=prunetree.constrained.time, model="BM", boot=100)
  ancova.bm3.r<-lm(rate ~ habitat + size + nodecounts, data=brmsdata)
  ancova.bm3.star<-phylolm::phylolm(rate ~ habitat + size + nodecounts, data=brmsdata, phy=starphylo, model="BM", boot=100)
  
  R2_lik(ancova.bm3, ancova.bm2) #partial R2 of size
  
  ancova.bm4<-phylolm::phylolm(rate ~ habitat + size + nodecounts + habitat:size, data=brmsdata, phy=prunetree.constrained.time, model="BM", boot=100)
  ancova.bm4.r<-lm(rate ~ habitat + size + nodecounts + habitat:size, data=brmsdata)
  ancova.bm4.star<-phylolm::phylolm(rate ~ habitat + size + nodecounts + habitat:size, data=brmsdata, phy=starphylo, model="BM", boot=100)
  
  R2_lik(ancova.bm4, ancova.bm3) #partial R2 of interaction
  
  #ancova.bm6<-phylolm::phylolm(rate~0 + habitat:size + size, data=brmsdata, phy=prunetree.shallow.time, model="BM", boot=100)
  bm.aic<-c(AICc.phylolm(ancova.bm1), AICc.phylolm(ancova.bm2), AICc.phylolm(ancova.bm3), AICc.phylolm(ancova.bm4))
  aic.w(bm.aic)
  
  #tmp<-teststep(formula = rate ~ habitat * size * nodecounts, data = brmsdata, phy=prunetree.shallow.time, model = "OUrandomRoot", direction="both")
  #tmp<-MASS:::stepAIC(object = lm(rate ~ habitat + size + nodecounts, data = brmsdata), direction="both")
  
  
  ##now with OU models
  
  ancova.ou1<-phylolm::phylolm(rate~ habitat, data=brmsdata, phy=prunetree.constrained.time, model="OUrandomRoot", boot=100)
  ancova.ou1.r<-lm(rate~ habitat, data=brmsdata)
  ancova.ou1.star<-phylolm::phylolm(rate~ habitat, data=brmsdata, phy=starphylo, model="OUrandomRoot", boot=100)
  
  R2_lik(ancova.ou1, ancova.ou1.star)
  
  ancova.ou2<-phylolm::phylolm(rate ~ habitat + nodecounts, data=brmsdata, phy=prunetree.constrained.time, model="OUrandomRoot", boot=100)
  ancova.ou2.r<-lm(rate ~ habitat + nodecounts, data=brmsdata)
  ancova.ou2.star<-phylolm::phylolm(rate ~ habitat + nodecounts, data=brmsdata, phy=starphylo, model="OUrandomRoot", boot=100)
  
  R2_lik(ancova.ou2, ancova.ou1) #partial R2 for nodecounts
  
  ancova.ou3<-phylolm::phylolm(rate ~ habitat + size + nodecounts, data=brmsdata, phy=prunetree.constrained.time, model="OUrandomRoot", boot=100)
  ancova.ou3.r<-lm(rate ~ habitat + size + nodecounts, data=brmsdata)
  ancova.ou3.star<-phylolm::phylolm(rate ~ habitat + size + nodecounts, data=brmsdata, phy=starphylo, model="OUrandomRoot", boot=100)
  
  R2_lik(ancova.ou3, ancova.ou2) #partial R2 of size
  
  ancova.ou4<-phylolm::phylolm(rate ~ habitat + size + nodecounts + habitat:size, data=brmsdata, phy=prunetree.constrained.time, model="OUrandomRoot", boot=100)
  ancova.ou4.r<-lm(rate ~ habitat + size + nodecounts + habitat:size, data=brmsdata)
  ancova.ou4.star<-phylolm::phylolm(rate ~ habitat + size + nodecounts + habitat:size, data=brmsdata, phy=starphylo, model="OUrandomRoot", boot=100)
  
  R2_lik(ancova.ou4, ancova.ou3) #partial R2 of interaction
  
  
  #with both BM and lambda, model with habitat is best
  ou.aic<-c(AICc.phylolm(ancova.ou1), AICc.phylolm(ancova.ou2), AICc.phylolm(ancova.ou3), AICc.phylolm(ancova.ou4))
  aic.w(ou.aic)
  
  all<-c(bm.aic, ou.aic)
  aic.w(all)
  
  #extract sig2 values
  bm.sig2s<-list()
  for(i in 1:4){
    bm.sig2s[i]<-eval(parse(text=paste("ancova.bm", i, sep="")))$sigma2
  }
  
  ou.sig2s<-list()
  for(i in 1:4){
    ou.sig2s[i]<-eval(parse(text=paste("ancova.ou", i, sep="")))$sigma2
  }
  
  ou.alpha<-list()
  for(i in 1:4){
    ou.alpha[i]<-eval(parse(text=paste("ancova.ou", i, sep="")))$optpar
  }
  
  require(rr2)
  bm.r2s<-rbind(R2(mod = ancova.bm1, mod.r=ancova.bm1.r, phy=prunetree.constrained.time),
                R2(mod = ancova.bm2, mod.r=ancova.bm2.r, phy=prunetree.constrained.time),
                R2(mod = ancova.bm3, mod.r=ancova.bm3.r, phy=prunetree.constrained.time),
                R2(mod = ancova.bm4, mod.r=ancova.bm4.r, phy=prunetree.constrained.time))
  # R2(mod = ancova.bm5, mod.r=ancova.bm5.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.bm6, mod.r=ancova.bm6.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.bm7, mod.r=ancova.bm7.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.bm8, mod.r=ancova.bm8.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.bm9, mod.r=ancova.bm9.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.bm10, mod.r=ancova.bm10.r, phy=prunetree.shallow.time))
  
  
  ou.r2s<-rbind(R2(mod = ancova.ou1, mod.r=ancova.ou1.r, phy=prunetree.constrained.time),
                R2(mod = ancova.ou2, mod.r=ancova.ou2.r, phy=prunetree.constrained.time),
                R2(mod = ancova.ou3, mod.r=ancova.ou3.r, phy=prunetree.constrained.time),
                R2(mod = ancova.ou4, mod.r=ancova.ou4.r, phy=prunetree.constrained.time))
  # R2(mod = ancova.ou5, mod.r=ancova.ou5.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.ou6, mod.r=ancova.ou6.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.ou7, mod.r=ancova.ou7.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.ou8, mod.r=ancova.ou8.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.ou9, mod.r=ancova.ou9.r, phy=prunetree.shallow.time),
  # R2(mod = ancova.ou10, mod.r=ancova.ou10.r, phy=prunetree.shallow.time))
  
  
  #summarize coefficients
  #https://stackoverflow.com/questions/29685806/printing-regression-coefficients-from-multiple-models-to-a-shared-data-frame
  # models<-list(ancova.bm1,
  #           ancova.bm2,
  #           ancova.bm3,
  #           ancova.bm4,
  #           ancova.bm5,
  #           ancova.bm6,
  #           ancova.bm7,
  #           ancova.bm8,
  #           ancova.bm9,
  #           ancova.bm10,
  #           ancova.ou1, 
  #           ancova.ou2, 
  #           ancova.ou3, 
  #           ancova.ou4, 
  #           ancova.ou5, 
  #           ancova.ou6, 
  #           ancova.ou7, 
  #           ancova.ou8, 
  #           ancova.ou9, 
  #           ancova.ou10)
  
  models<-list(ancova.bm1,
               ancova.bm2,
               ancova.bm3,
               ancova.bm4,
               ancova.ou1, 
               ancova.ou2, 
               ancova.ou3, 
               ancova.ou4)
  
  names(models)<-paste0("MODEL", 1:8)
  
  Pvals<-list()
  for(i in 1:8){
    Pvals[[i]]<-summary(models[[i]])$coefficients[,6]
  }
  
  
  for(i in 1:length(Pvals)){
    for(j in 1:length(Pvals[[i]])){
      swap<-if(Pvals[[i]][[j]] < 0.05){paste("*")}else{paste("")}
      Pvals[[i]][[j]]<-swap
    }
  }
  
  coefs<-lapply(models, coefficients)
  coefs<-lapply(coefs, round, 4)
  
  for(i in 1:length(coefs)){
    for(j in 1:length(coefs[[i]])){
      coefs[[i]][[j]] <- paste(coefs[[i]][[j]], Pvals[[i]][[j]], sep="")
    }
  }
  
  
  coefs<-lapply(coefs, unlist)
  coefs<-lapply(coefs, as.list)
  coefs<-data.table::rbindlist(coefs, fill=T, use.names=T)
  
  
  summary<-data.frame(process=c(rep("BM", 4),rep("OU", 4)), 
                      #vars=c(forms,forms), 
                      AICc=round(c(bm.aic, ou.aic), 2), 
                      weight=as.numeric(round(aic.w(all),2)), 
                      alpha=round(c(rep(NA, 4), unlist(ou.alpha)),4), 
                      sig2=round(unlist(c(bm.sig2s, ou.sig2s)), 4), 
                      R2pred=rbind(bm.r2s, ou.r2s)[,3], 
                      coefs)
  
  
  summary<-summary[order(summary$weight, decreasing =T),]
  rownames(summary)<-NULL
  
  
  require(stargazer)
  stargazer(summary, summary=F, type="latex")
  
}


#plot(rate ~ habitat, data=brmsdata)
require(ggthemes)
require(ggplot2)
require(ggpubr)

pglsdata$habitat<-ordered(pglsdata$habitat, levels=c('OF', 'Vent', 'Seep'))
#boxplots for constrained
#phyl.aov.constrained
# {
# pp <- ggplot(pglsdata, aes(habitat, r2t.log)) + geom_boxplot() #+
# #scale_fill_manual(breaks = habitat,
# #                values = c("green", "black", "black"))
# 
# #df1 <- data.frame(a = c(1, 1:3,3), b = c(0.69, 0.71, 0.71, 0.71, 0.69)-0.4)
# #df2 <- data.frame(a = c(1, 1,2, 2), b = c(0.25, 0.27, 0.27, 0.25)-0.075)
# #df3 <- data.frame(a = c(2, 2, 3, 3), b = c(0.25, 0.27, 0.27, 0.25)-0.2)
# 
# pdf(file="tubeworm_boxplots-constrained.pdf", width=5, height=5)
# pp + #geom_line(data = df1, aes(x = a, y = b)) + 
#   #annotate("text", x = 2, y = 0.33, label = "0.02", size = 4) +
#   #geom_line(data = df2, aes(x = a, y = b)) + 
#   #annotate("text", x = 1.5, y = 0.22, label = "0.03", size = 4) +
#   #geom_line(data = df3, aes(x = a, y = b)) + 
#   #annotate("text", x = 2.5, y = 0.1, label = "0.45", size = 4) +
#   theme_clean()
# 
# dev.off()
# }

{
  pp <-
    ggplot(pglsdata, aes(habitat, r2t.log)) + geom_boxplot(
      aes(fill=habitat),
      outlier.shape = NA,
      width = 0.35,
      position = position_nudge(x = 0.175),
      show.legend = FALSE
    )
  

  pdf(file="tubeworm_boxplots-constrained.pdf", width=5, height=5)
  pp + 
    #theme_clean() + geom_jitter(data=data.frame(size=log(pglsdata[,c(5)]),habitat=pglsdata[,c(3)]), width=0.1)
    theme_classic() + geom_point(color='black', fill='black',
      shape = 21,
      size = 1.5,
      #aes(fill = as.character(pglsdata$pch)),
      data = data.frame(r2t.log = pglsdata$r2t.log, habitat = pglsdata$habitat),
      show.legend = FALSE,
      position = sdamr::position_jitternudge(jitter.width = 0.2, nudge.x = -0.175)
    ) +
    scale_fill_manual(values = c(
      #"21" = scales::alpha("black", 0.9),
      #"23" = scales::alpha("white", 0.9),
      "OF" = "#007AC1",
      'Vent' = "#00BCDF",
      "Seep" = "#CAE3FC"
    ))
  
  dev.off()
}


#boxplots for constrained
#phyl.aov.unconstrained
# {
#   pp <- ggplot(pglsdata, aes(habitat, r2t.unconstrained.log)) + geom_boxplot() #+
#   #scale_fill_manual(breaks = habitat,
#   #                values = c("green", "black", "black"))
#   
#   df1 <- data.frame(a = c(1, 1:3,3), b = c(0.69, 0.71, 0.71, 0.71, 0.69)-0.4)
#   df2 <- data.frame(a = c(1, 1,2, 2), b = c(0.25, 0.27, 0.27, 0.25)-0.075)
#   df3 <- data.frame(a = c(2, 2, 3, 3), b = c(0.25, 0.27, 0.27, 0.25)-0.2)
#   
#   pdf(file="tubeworm_boxplots-unconstrained.pdf", width=5, height=5)
#   pp + geom_line(data = df1, aes(x = a, y = b)) + 
#     annotate("text", x = 2, y = 0.33, label = "0.05", size = 4) +
#     geom_line(data = df2, aes(x = a, y = b)) + 
#     annotate("text", x = 1.5, y = 0.22, label = "0.09", size = 4) +
#     geom_line(data = df3, aes(x = a, y = b)) + 
#     annotate("text", x = 2.5, y = 0.1, label = "0.33", size = 4) +
#     theme_clean()
#   
#   dev.off()
# }


{
  pp <-
    ggplot(pglsdata, aes(habitat, r2t.unconstrained.log)) + geom_boxplot(
      aes(fill=habitat),
      outlier.shape = NA,
      width = 0.35,
      position = position_nudge(x = 0.175),
      show.legend = FALSE
    )
  
  
  df1 <- data.frame(a = c(1, 1:3,3), b = c(0.69, 0.71, 0.71, 0.71, 0.69)-0.4)
  df2 <- data.frame(a = c(1, 1,2, 2), b = c(0.25, 0.27, 0.27, 0.25)-0.075)
  df3 <- data.frame(a = c(2, 2, 3, 3), b = c(0.25, 0.27, 0.27, 0.25)-0.2)
  
  
  pdf(file="tubeworm_boxplots-unconstrained.pdf", width=5, height=5)
  pp + 
    geom_line(data = df1, aes(x = a, y = b)) + 
    annotate("text", x = 2, y = 0.33, label = "0.05", size = 4) +
    geom_line(data = df2, aes(x = a, y = b)) + 
    annotate("text", x = 1.5, y = 0.22, label = "0.095", size = 4) +
    geom_line(data = df3, aes(x = a, y = b)) + 
    annotate("text", x = 2.5, y = 0.1, label = "0.33", size = 4) +
    #theme_clean() + geom_jitter(data=data.frame(size=log(pglsdata[,c(5)]),habitat=pglsdata[,c(3)]), width=0.1)
    theme_classic() + geom_point(color='black', fill='black',
                                 shape = 21,
                                 size = 1.5,
                                 #aes(fill = as.character(pglsdata$pch)),
                                 data = data.frame(r2t.unconstrained.log = pglsdata$r2t.unconstrained.log, habitat = pglsdata$habitat),
                                 show.legend = FALSE,
                                 position = sdamr::position_jitternudge(jitter.width = 0.2, nudge.x = -0.175)
    ) +
    scale_fill_manual(values = c(
      #"21" = scales::alpha("black", 0.9),
      #"23" = scales::alpha("white", 0.9),
      "OF" = "#007AC1",
      'Vent' = "#00BCDF",
      "Seep" = "#CAE3FC"
    ))
  
  dev.off()
}


phyl.aov.size<-phylANOVA(tree=prunetree.constrained.time, x=setNames(brmsdata$habitat, rownames(brmsdata)), y= setNames(brmsdata$size, rownames(brmsdata)), nsim = 1000)
brmsdata$habitat<-factor(brmsdata$habitat, levels=c("OF", "Vent", "Seep"))
#no significant differences in size after accounting for phylogeny
# ANOVA table: Phylogenetic ANOVA
# 
# Response: y
# Sum Sq   Mean Sq   F value Pr(>F)
# x        166.7001 83.350070 58.002073  0.184
# Residual 102.0283  1.437019                 
# 
# P-value based on simulation.
# ---------
#   
#   Pairwise posthoc test using method = "holm"
# 
# Pairwise t-values:
#   OF      Seep      Vent
# OF   0.000000 -9.103291 -8.029208
# Seep 9.103291  0.000000  1.536035
# Vent 8.029208 -1.536035  0.000000
# 
# Pairwise corrected P-values:
#   OF  Seep  Vent
# OF   1.000 0.396 0.466
# Seep 0.396 1.000 0.466
# Vent 0.466 0.466 1.000
# ---------


{
  pp <-
    ggplot(brmsdata, aes(habitat, size)) + geom_boxplot(
      aes(fill=habitat),
      outlier.shape = NA,
      width = 0.35,
      position = position_nudge(x = 0.175),
      show.legend = FALSE
    )
  
  df1 <- data.frame(a = c(1, 1:3,3), b = c(0.69-0.1, 0.71, 0.71, 0.71, 0.69-0.1)+10.3)
  df2 <- data.frame(a = c(1, 1,2, 2), b = c(0.25-0.1, 0.27, 0.27, 0.25-0.1)+9.5)
  df3 <- data.frame(a = c(2, 2, 3, 3), b = c(0.25-0.1, 0.27, 0.27, 0.25-0.1)+8.5)
  
  pdf(file="tubeworm_no_fren_boxplots_size_habitat.pdf", width=5, height=5)
  pp + #geom_line(data = df1, aes(x = a, y = b)) + 
    #annotate("text", x = 2, y = 11.5, label = "0.425", size = 4) +
    #geom_line(data = df2, aes(x = a, y = b)) + 
    #annotate("text", x = 1.5, y = 10.2, label = "0.405", size = 4) +
    #geom_line(data = df3, aes(x = a, y = b)) + 
    #annotate("text", x = 2.5, y = 9.3, label = "0.425", size = 4) +
    #theme_clean() + geom_jitter(data=data.frame(size=log(pglsdata[,c(5)]),habitat=pglsdata[,c(3)]), width=0.1)
    theme_classic() + geom_point(
      shape = 21,
      size = 1.5,
      aes(fill = as.character(pglsdata$pch)),
      data = data.frame(size = brmsdata$size, habitat = brmsdata$habitat),
      show.legend = FALSE,
      position = sdamr::position_jitternudge(jitter.width = 0.2, nudge.x = -0.175)
    ) +
    scale_fill_manual(values = c(
      "21" = scales::alpha("black", 0.9),
      "23" = scales::alpha("white", 0.9),
      "OF" = "#007AC1",
      'Vent' = "#00BCDF",
      "Seep" = "#CAE3FC"
    ))
  
  dev.off()
}

#write output for amelia
write.table(pglsdata, file='data.csv', quote = F, sep = ",")

## checking variances of bootstrap trees
fit<-lm(log(compute_summary_stats(CO1_ML.constrained.boottrees)$Mean)~log(compute_summary_stats(CO1_ML.constrained)$Mean))
plot(log(compute_summary_stats(CO1_ML.constrained.boottrees)$Mean)~log(compute_summary_stats(CO1_ML.constrained)$Mean))
summary(fit)

fit<-lm(log(compute_summary_stats(CO1_ML.unconstrained.boottrees)$Mean)~log(compute_summary_stats(CO1_ML.unconstrained)$Mean))
plot(log(compute_summary_stats(CO1_ML.unconstrained.boottrees)$Mean)~log(compute_summary_stats(CO1_ML.unconstrained)$Mean))
summary(fit)

median(compute_summary_stats(CO1_ML.constrained.boottrees)$CV)


